diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index 27c3026..4c58b4d 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,5 +1,9 @@
 // $Id$
 
+Drupal 6.8, 2008-12-11
+----------------------
+- Removed a previous change incompatible with PHP 5.1.x and lower.
+
 Drupal 6.7, 2008-12-10
 ----------------------
 - Fixed security issues, (Cross site request forgery and Cross site scripting), see SA-2008-073
@@ -150,6 +154,10 @@ Drupal 6.0, 2008-02-13
 - Removed old system updates. Updates from Drupal versions prior to 5.x will
   require upgrading to 5.x before upgrading to 6.x.
 
+Drupal 5.14, 2008-12-11
+----------------------
+- Removed a previous change incompatible with PHP 5.1.x and lower.
+
 Drupal 5.13, 2008-12-10
 -----------------------
 - fixed a variety of small bugs.
diff --git a/includes/session.inc b/includes/session.inc
index bf730e5..725c45d 100644
--- a/includes/session.inc
+++ b/includes/session.inc
@@ -97,10 +97,6 @@ function sess_regenerate() {
     setcookie(session_name(), '', time() - 42000, '/');
   }
 
-  extract(session_get_cookie_params());
-  // Set "httponly" to TRUE to reduce the risk of session stealing via XSS.
-  // This has no effect for PHP < 5.2.0.
-  session_set_cookie_params($lifetime, $path, $domain, $secure, TRUE);
   session_regenerate_id();
 
   db_query("UPDATE {sessions} SET sid = '%s' WHERE sid = '%s'", session_id(), $old_session_id);
diff --git a/modules/system/system.module b/modules/system/system.module
index d3673c3..9ff9312 100644
--- a/modules/system/system.module
+++ b/modules/system/system.module
@@ -9,7 +9,7 @@
 /**
  * The current system version.
  */
-define('VERSION', '6.7');
+define('VERSION', '6.8');
 
 /**
  * Core API compatibility.
