diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index 794cdb9..69d02fc 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,4 +1,12 @@
-// $Id: CHANGELOG.txt,v 1.253.2.22 2009/01/14 23:34:06 goba Exp $
+// $Id: CHANGELOG.txt,v 1.253.2.25 2009/02/25 21:02:36 goba Exp $
+
+Drupal 6.10, 2009-02-25
+----------------------
+- Fixed a security issue, (Local file inclusion on Windows),
+  see SA-CORE-2009-003
+- Fixed node_feed() so custom fields can show up in RSS feeds.
+- Improved PostgreSQL compatibility.
+- Fixed a variety of small bugs.
 
 Drupal 6.9, 2009-01-14
 ----------------------
@@ -11,9 +19,6 @@ Drupal 6.9, 2009-01-14
   changed for contributed modules, see http://drupal.org/node/322731.
 - Fixed a variety of small bugs.
 
-Drupal 6.9-dev, xxxx-xx-xx (development release)
-----------------------
-
 Drupal 6.8, 2008-12-11
 ----------------------
 - Removed a previous change incompatible with PHP 5.1.x and lower.
diff --git a/includes/actions.inc b/includes/actions.inc
index 946c939..3b22ac2 100644
--- a/includes/actions.inc
+++ b/includes/actions.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: actions.inc,v 1.8.2.6 2009/01/06 17:02:58 goba Exp $
+// $Id: actions.inc,v 1.8.2.7 2009/02/16 14:34:30 goba Exp $
 
 /**
  * @file
@@ -40,6 +40,7 @@
  *   performs the action, keyed on action ID.
  */
 function actions_do($action_ids, &$object, $context = NULL, $a1 = NULL, $a2 = NULL) {
+  // $stack tracks the number of recursive calls.
   static $stack;
   $stack++;
   if ($stack > variable_get('actions_max_stack', 35)) {
@@ -103,6 +104,7 @@ function actions_do($action_ids, &$object, $context = NULL, $a1 = NULL, $a2 = NU
       $result[$action_ids] = $action_ids($object, $context, $a1, $a2);
     }
   }
+  $stack--;
   return $result;
 }
 
diff --git a/includes/bootstrap.inc b/includes/bootstrap.inc
index d1304e4..70e5156 100644
--- a/includes/bootstrap.inc
+++ b/includes/bootstrap.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: bootstrap.inc,v 1.206.2.9 2009/01/14 19:10:25 goba Exp $
+// $Id: bootstrap.inc,v 1.206.2.11 2009/02/25 13:49:54 dries Exp $
 
 /**
  * @file
@@ -132,6 +132,16 @@ define('LANGUAGE_NEGOTIATION_PATH', 2);
 define('LANGUAGE_NEGOTIATION_DOMAIN', 3);
 
 /**
+ * Language written left to right. Possible value of $language->direction.
+ */
+define('LANGUAGE_LTR', 0);
+
+/**
+ * Language written right to left. Possible value of $language->direction.
+ */
+define('LANGUAGE_RTL', 1);
+
+/**
  * Start the timer with the specified name. If you start and stop
  * the same timer multiple times, the measured intervals will be
  * accumulated.
@@ -551,7 +561,7 @@ function page_get_cache($status_only = FALSE) {
   }
   $cache = NULL;
 
-  if (!$user->uid && $_SERVER['REQUEST_METHOD'] == 'GET' && count(drupal_set_message()) == 0) {
+  if (!$user->uid && $_SERVER['REQUEST_METHOD'] == 'GET' && count(drupal_set_message()) == 0 && $_SERVER['SERVER_SOFTWARE'] !== 'PHP CLI') {
     $cache = cache_get($base_root . request_uri(), 'cache_page');
 
     if (empty($cache)) {
diff --git a/includes/common.inc b/includes/common.inc
index 2f1e865..8d240db 100644
--- a/includes/common.inc
+++ b/includes/common.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: common.inc,v 1.756.2.42 2009/01/14 23:34:07 goba Exp $
+// $Id: common.inc,v 1.756.2.47 2009/02/25 21:02:36 goba Exp $
 
 /**
  * @file
@@ -1411,7 +1411,6 @@ function url($path = NULL, $options = array()) {
 
   global $base_url;
   static $script;
-  static $clean_url;
 
   if (!isset($script)) {
     // On some web servers, such as IIS, we can't omit "index.php". So, we
@@ -1420,11 +1419,6 @@ function url($path = NULL, $options = array()) {
     $script = (strpos($_SERVER['SERVER_SOFTWARE'], 'Apache') === FALSE) ? 'index.php' : '';
   }
 
-  // Cache the clean_url variable to improve performance.
-  if (!isset($clean_url)) {
-    $clean_url = (bool)variable_get('clean_url', '0');
-  }
-
   if (!isset($options['base_url'])) {
     // The base_url might be rewritten from the language rewrite in domain mode.
     $options['base_url'] = $base_url;
@@ -1450,7 +1444,7 @@ function url($path = NULL, $options = array()) {
   $prefix = empty($path) ? rtrim($options['prefix'], '/') : $options['prefix'];
   $path = drupal_urlencode($prefix . $path);
 
-  if ($clean_url) {
+  if (variable_get('clean_url', '0')) {
     // With Clean URLs.
     if ($options['query']) {
       return $base . $path .'?'. $options['query'] . $options['fragment'];
@@ -1539,6 +1533,8 @@ function drupal_attributes($attributes = array()) {
  *   an HTML string containing a link to the given path.
  */
 function l($text, $path, $options = array()) {
+  global $language;
+
   // Merge in defaults.
   $options += array(
       'attributes' => array(),
@@ -1546,7 +1542,8 @@ function l($text, $path, $options = array()) {
     );
 
   // Append active class.
-  if ($path == $_GET['q'] || ($path == '<front>' && drupal_is_front_page())) {
+  if (($path == $_GET['q'] || ($path == '<front>' && drupal_is_front_page())) &&
+      (empty($options['language']) || $options['language']->language == $language->language)) {
     if (isset($options['attributes']['class'])) {
       $options['attributes']['class'] .= ' active';
     }
@@ -1757,7 +1754,7 @@ function drupal_add_css($path = NULL, $type = 'module', $media = 'all', $preproc
     $css[$media][$type][$path] = $preprocess;
 
     // If the current language is RTL, add the CSS file with RTL overrides.
-    if (defined('LANGUAGE_RTL') && $language->direction == LANGUAGE_RTL) {
+    if ($language->direction == LANGUAGE_RTL) {
       $rtl_path = str_replace('.css', '-rtl.css', $path);
       if (file_exists($rtl_path)) {
         $css[$media][$type][$rtl_path] = $preprocess;
diff --git a/includes/database.inc b/includes/database.inc
index d017a84..5e70c2d 100644
--- a/includes/database.inc
+++ b/includes/database.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: database.inc,v 1.92.2.3 2008/10/20 09:13:04 goba Exp $
+// $Id: database.inc,v 1.92.2.4 2009/02/16 14:41:58 goba Exp $
 
 /**
  * @file
@@ -458,7 +458,7 @@ function db_escape_table($string) {
  *
  *  - 'primary key': An array of one or more key column specifiers (see below)
  *    that form the primary key.
- *  - 'unique key': An associative array of unique keys ('keyname' =>
+ *  - 'unique keys': An associative array of unique keys ('keyname' =>
  *    specification).  Each specification is an array of one or more
  *    key column specifiers (see below) that form a unique key on the table.
  *  - 'indexes':  An associative array of indexes ('indexame' =>
diff --git a/includes/form.inc b/includes/form.inc
index 4da2390..fb1691e 100644
--- a/includes/form.inc
+++ b/includes/form.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: form.inc,v 1.265.2.17 2009/01/12 16:07:23 goba Exp $
+// $Id: form.inc,v 1.265.2.19 2009/02/22 18:12:46 goba Exp $
 
 /**
  * @defgroup forms Form builder functions
@@ -1169,7 +1169,12 @@ function form_type_image_button_value($form, $edit = FALSE) {
  */
 function form_type_checkbox_value($form, $edit = FALSE) {
   if ($edit !== FALSE) {
-    return !empty($edit) ? $form['#return_value'] : 0;
+    if (empty($form['#disabled'])) {
+      return !empty($edit) ? $form['#return_value'] : 0;
+    }
+    else {
+      return $form['#default_value'];
+    }
   }
 }
 
@@ -1533,7 +1538,7 @@ function theme_radio($element) {
   $output .= (check_plain($element['#value']) == $element['#return_value']) ? ' checked="checked" ' : ' ';
   $output .= drupal_attributes($element['#attributes']) .' />';
   if (!is_null($element['#title'])) {
-    $output = '<label class="option">'. $output .' '. $element['#title'] .'</label>';
+    $output = '<label class="option" for="'. $element['#id'] .'">'. $output .' '. $element['#title'] .'</label>';
   }
 
   unset($element['#title']);
@@ -1872,7 +1877,7 @@ function theme_checkbox($element) {
   $checkbox .= drupal_attributes($element['#attributes']) .' />';
 
   if (!is_null($element['#title'])) {
-    $checkbox = '<label class="option">'. $checkbox .' '. $element['#title'] .'</label>';
+    $checkbox = '<label class="option" for="'. $element['#id'] .'">'. $checkbox .' '. $element['#title'] .'</label>';
   }
 
   unset($element['#title']);
diff --git a/includes/install.inc b/includes/install.inc
index 39892d6..5ace0bb 100644
--- a/includes/install.inc
+++ b/includes/install.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: install.inc,v 1.56.2.3 2008/08/14 09:26:43 dries Exp $
+// $Id: install.inc,v 1.56.2.4 2009/02/16 10:25:02 goba Exp $
 
 define('SCHEMA_UNINSTALLED', -1);
 define('SCHEMA_INSTALLED', 0);
@@ -281,7 +281,7 @@ function drupal_verify_profile($profile, $locale) {
 
   // Get a list of modules required by this profile.
   $function = $profile .'_profile_modules';
-  $module_list = array_merge(drupal_required_modules(), $function(), ($locale != 'en' ? array('locale') : array()));
+  $module_list = array_merge(drupal_required_modules(), $function(), ($locale != 'en' && !empty($locale) ? array('locale') : array()));
 
   // Get a list of modules that exist in Drupal's assorted subdirectories.
   $present_modules = array();
diff --git a/includes/menu.inc b/includes/menu.inc
index eb95127..29635e6 100644
--- a/includes/menu.inc
+++ b/includes/menu.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: menu.inc,v 1.255.2.27 2008/11/10 17:30:39 goba Exp $
+// $Id: menu.inc,v 1.255.2.28 2009/02/09 16:28:21 dries Exp $
 
 /**
  * @file
@@ -552,15 +552,18 @@ function _menu_item_localize(&$item, $map, $link_translate = FALSE) {
  *   a non existing node) then this function return FALSE.
  */
 function _menu_translate(&$router_item, $map, $to_arg = FALSE) {
+  if ($to_arg) {
+    // Fill in missing path elements, such as the current uid.
+    _menu_link_map_translate($map, $router_item['to_arg_functions']);
+  }
+  // The $path_map saves the pieces of the path as strings, while elements in
+  // $map may be replaced with loaded objects.
   $path_map = $map;
   if (!_menu_load_objects($router_item, $map)) {
     // An error occurred loading an object.
     $router_item['access'] = FALSE;
     return FALSE;
   }
-  if ($to_arg) {
-    _menu_link_map_translate($path_map, $router_item['to_arg_functions']);
-  }
 
   // Generate the link path for the page request or local tasks.
   $link_map = explode('/', $router_item['path']);
diff --git a/includes/module.inc b/includes/module.inc
index 22a8466..33bee08 100644
--- a/includes/module.inc
+++ b/includes/module.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: module.inc,v 1.115 2007/12/27 12:31:05 goba Exp $
+// $Id: module.inc,v 1.115.2.1 2009/02/16 10:32:10 goba Exp $
 
 /**
  * @file
@@ -48,8 +48,8 @@ function module_list($refresh = FALSE, $bootstrap = TRUE, $sort = FALSE, $fixed_
   static $list, $sorted_list;
 
   if ($refresh || $fixed_list) {
-    unset($sorted_list);
     $list = array();
+    $sorted_list = NULL;
     if ($fixed_list) {
       foreach ($fixed_list as $name => $module) {
         drupal_get_filename('module', $name, $module['filename']);
diff --git a/includes/theme.inc b/includes/theme.inc
index 430c6ab..1abe305 100644
--- a/includes/theme.inc
+++ b/includes/theme.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: theme.inc,v 1.415.2.17 2008/12/10 21:09:05 goba Exp $
+// $Id: theme.inc,v 1.415.2.19 2009/02/25 21:02:16 goba Exp $
 
 /**
  * @file
@@ -1133,6 +1133,7 @@ function theme_status_messages($display = NULL) {
  *   A string containing an unordered list of links.
  */
 function theme_links($links, $attributes = array('class' => 'links')) {
+  global $language;
   $output = '';
 
   if (count($links) > 0) {
@@ -1151,7 +1152,8 @@ function theme_links($links, $attributes = array('class' => 'links')) {
       if ($i == $num_links) {
         $class .= ' last';
       }
-      if (isset($link['href']) && ($link['href'] == $_GET['q'] || ($link['href'] == '<front>' && drupal_is_front_page()))) {
+      if (isset($link['href']) && ($link['href'] == $_GET['q'] || ($link['href'] == '<front>' && drupal_is_front_page()))
+          && (empty($link['language']) || $link['language']->language == $language->language)) {
         $class .= ' active';
       }
       $output .= '<li'. drupal_attributes(array('class' => $class)) .'>';
@@ -1866,6 +1868,7 @@ function template_preprocess_page(&$variables) {
   $suggestion = 'page';
   $suggestions = array();
   while ($arg = arg($i++)) {
+    $arg = str_replace(array('/', '\\', '\0'), '', $arg);
     $suggestions[] = $suggestion .'-'. $arg;
     if (!is_numeric($arg)) {
       $suggestion .= '-'. $arg;
diff --git a/install.php b/install.php
index bb9fb0b..a13f8b8 100644
--- a/install.php
+++ b/install.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: install.php,v 1.113.2.7 2008/10/22 16:31:37 goba Exp $
+// $Id: install.php,v 1.113.2.8 2009/02/25 11:47:36 goba Exp $
 
 require_once './includes/install.inc';
 
@@ -556,9 +556,11 @@ function install_select_locale($profilename) {
       }
     }
 
-    foreach ($locales as $locale) {
-      if ($_POST['locale'] == $locale->name) {
-        return $locale->name;
+    if (!empty($_POST['locale'])) {
+      foreach ($locales as $locale) {
+        if ($_POST['locale'] == $locale->name) {
+          return $locale->name;
+        }
       }
     }
 
diff --git a/modules/aggregator/aggregator.info b/modules/aggregator/aggregator.info
index 5cd8b54..053055d 100644
--- a/modules/aggregator/aggregator.info
+++ b/modules/aggregator/aggregator.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/block/block.info b/modules/block/block.info
index 2a4e1d6..527c728 100644
--- a/modules/block/block.info
+++ b/modules/block/block.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/blog/blog.info b/modules/blog/blog.info
index 751cd43..86c8eb3 100644
--- a/modules/blog/blog.info
+++ b/modules/blog/blog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/blog/blog.module b/modules/blog/blog.module
index e11ca18..6e7d998 100644
--- a/modules/blog/blog.module
+++ b/modules/blog/blog.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: blog.module,v 1.297.2.3 2008/05/19 07:27:35 goba Exp $
+// $Id: blog.module,v 1.297.2.4 2009/02/25 12:21:53 goba Exp $
 
 /**
  * @file
@@ -49,7 +49,8 @@ function blog_user($type, &$edit, &$user) {
     $user->content['summary']['blog'] =  array(
       '#type' => 'user_profile_item',
       '#title' => t('Blog'),
-      '#value' => l(t('View recent blog entries'), "blog/$user->uid", array('attributes' => array('title' => t("Read @username's latest blog entries.", array('@username' => $user->name))))),
+      // l() escapes the attributes, so we should not escape !username here.
+      '#value' => l(t('View recent blog entries'), "blog/$user->uid", array('attributes' => array('title' => t("Read !username's latest blog entries.", array('!username' => $user->name))))),
       '#attributes' => array('class' => 'blog'),
     );
   }
@@ -104,8 +105,8 @@ function blog_form(&$node) {
  */
 function blog_view($node, $teaser = FALSE, $page = FALSE) {
   if ($page) {
-    // Breadcrumb navigation
-    drupal_set_breadcrumb(array(l(t('Home'), NULL), l(t('Blogs'), 'blog'), l(t("@name's blog", array('@name' => $node->name)), 'blog/'. $node->uid)));
+    // Breadcrumb navigation. l() escapes the title, so we should not escape !name. 
+    drupal_set_breadcrumb(array(l(t('Home'), NULL), l(t('Blogs'), 'blog'), l(t("!name's blog", array('!name' => $node->name)), 'blog/'. $node->uid)));
   }
   return node_prepare($node, $teaser);
 }
@@ -118,10 +119,11 @@ function blog_link($type, $node = NULL, $teaser = FALSE) {
 
   if ($type == 'node' && $node->type == 'blog') {
     if (arg(0) != 'blog' || arg(1) != $node->uid) {
+      // This goes to l() and therefore escapes !username in both the title and attributes.
       $links['blog_usernames_blog'] = array(
-        'title' => t("@username's blog", array('@username' => $node->name)),
+        'title' => t("!username's blog", array('!username' => $node->name)),
         'href' => "blog/$node->uid",
-        'attributes' => array('title' => t("Read @username's latest blog entries.", array('@username' => $node->name)))
+        'attributes' => array('title' => t("Read !username's latest blog entries.", array('!username' => $node->name)))
       );
     }
   }
diff --git a/modules/blogapi/blogapi.info b/modules/blogapi/blogapi.info
index 5ebf168..b62897d 100644
--- a/modules/blogapi/blogapi.info
+++ b/modules/blogapi/blogapi.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/book/book.info b/modules/book/book.info
index 4f36642..e6e7815 100644
--- a/modules/book/book.info
+++ b/modules/book/book.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/book/book.module b/modules/book/book.module
index 6fa42b2..9389b2f 100644
--- a/modules/book/book.module
+++ b/modules/book/book.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: book.module,v 1.454.2.5 2008/07/03 06:38:46 dries Exp $
+// $Id: book.module,v 1.454.2.6 2009/02/25 11:47:37 goba Exp $
 
 /**
  * @file
@@ -885,7 +885,7 @@ function template_preprocess_book_export_html(&$variables) {
   $variables['title'] = check_plain($variables['title']);
   $variables['base_url'] = $base_url;
   $variables['language'] = $language;
-  $variables['language_rtl'] = (defined('LANGUAGE_RTL') && $language->direction == LANGUAGE_RTL) ? TRUE : FALSE;
+  $variables['language_rtl'] = ($language->direction == LANGUAGE_RTL);
   $variables['head'] = drupal_get_html_head();
 }
 
diff --git a/modules/color/color.info b/modules/color/color.info
index c9f77f0..6b15d1b 100644
--- a/modules/color/color.info
+++ b/modules/color/color.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/color/color.module b/modules/color/color.module
index db73768..045fa45 100644
--- a/modules/color/color.module
+++ b/modules/color/color.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: color.module,v 1.39 2008/01/23 09:43:25 goba Exp $
+// $Id: color.module,v 1.39.2.1 2009/02/25 11:47:37 goba Exp $
 
 /**
  * Implementation of hook_help().
@@ -93,7 +93,7 @@ function _color_page_alter(&$vars) {
 
           // If the current language is RTL and the CSS file had an RTL variant,
           // pull out the non-colored and add rewritten RTL stylesheet.
-          if (defined('LANGUAGE_RTL') && $language->direction == LANGUAGE_RTL) {
+          if ($language->direction == LANGUAGE_RTL) {
             $rtl_old_path = str_replace('.css', '-rtl.css', $old_path);
             $rtl_color_path = str_replace('.css', '-rtl.css', $color_path);
             if (file_exists($rtl_color_path)) {
diff --git a/modules/comment/comment.info b/modules/comment/comment.info
index 3ab7e8b..6ff2108 100644
--- a/modules/comment/comment.info
+++ b/modules/comment/comment.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/contact/contact.info b/modules/contact/contact.info
index f96c492..1b7f2da 100644
--- a/modules/contact/contact.info
+++ b/modules/contact/contact.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/dblog/dblog.info b/modules/dblog/dblog.info
index 3ef82b1..5c31697 100644
--- a/modules/dblog/dblog.info
+++ b/modules/dblog/dblog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/filter/filter.info b/modules/filter/filter.info
index 4795370..19b6987 100644
--- a/modules/filter/filter.info
+++ b/modules/filter/filter.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/forum/forum.info b/modules/forum/forum.info
index 4bf30b2..c450515 100644
--- a/modules/forum/forum.info
+++ b/modules/forum/forum.info
@@ -7,8 +7,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/help/help.info b/modules/help/help.info
index a22ac79..1d237b4 100644
--- a/modules/help/help.info
+++ b/modules/help/help.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/locale/locale.info b/modules/locale/locale.info
index ab23679..0efb98e 100644
--- a/modules/locale/locale.info
+++ b/modules/locale/locale.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/locale/locale.module b/modules/locale/locale.module
index 5ecf0ba..1ad7576 100644
--- a/modules/locale/locale.module
+++ b/modules/locale/locale.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: locale.module,v 1.212.2.5 2008/10/20 09:31:06 goba Exp $
+// $Id: locale.module,v 1.212.2.6 2009/02/25 11:47:37 goba Exp $
 
 /**
  * @file
@@ -12,17 +12,6 @@
  *   Gettext portable object files are supported.
  */
 
-/**
- * Language written left to right. Possible value of $language->direction.
- */
-define('LANGUAGE_LTR', 0);
-
-/**
- * Language written right to left. Possible value of $language->direction.
- */
-define('LANGUAGE_RTL', 1);
-
-
 // ---------------------------------------------------------------------------------
 // Hook implementations
 
diff --git a/modules/menu/menu.admin.inc b/modules/menu/menu.admin.inc
index b49c4d7..f26e840 100644
--- a/modules/menu/menu.admin.inc
+++ b/modules/menu/menu.admin.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: menu.admin.inc,v 1.26.2.3 2008/02/11 15:12:53 goba Exp $
+// $Id: menu.admin.inc,v 1.26.2.4 2009/02/25 13:15:40 goba Exp $
 
 /**
  * @file
@@ -366,7 +366,7 @@ function menu_item_delete_submit($form, &$form_state) {
  * Process menu and menu item add/edit form submissions.
  */
 function menu_edit_item_submit($form, &$form_state) {
-  $item = $form_state['values']['menu'];
+  $item = &$form_state['values']['menu'];
 
   // The value of "hidden" is the opposite of the value
   // supplied by the "enabled" checkbox.
diff --git a/modules/menu/menu.info b/modules/menu/menu.info
index dac1748..4060db1 100644
--- a/modules/menu/menu.info
+++ b/modules/menu/menu.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/menu/menu.module b/modules/menu/menu.module
index 54f9801..497abbd 100644
--- a/modules/menu/menu.module
+++ b/modules/menu/menu.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: menu.module,v 1.157.2.3 2008/05/10 06:53:53 dries Exp $
+// $Id: menu.module,v 1.157.2.5 2009/02/25 13:15:40 goba Exp $
 
 /**
  * @file
@@ -163,11 +163,13 @@ function menu_theme() {
  */
 function menu_enable() {
   menu_rebuild();
-  $link = db_fetch_array(db_query("SELECT mlid AS plid, menu_name from {menu_links} WHERE link_path = 'admin/build/menu' AND module = 'system'"));
-  $link['router_path'] = 'admin/build/menu-customize/%';
-  $link['module'] = 'menu';
+  $base_link = db_fetch_array(db_query("SELECT mlid AS plid, menu_name from {menu_links} WHERE link_path = 'admin/build/menu' AND module = 'system'"));
+  $base_link['router_path'] = 'admin/build/menu-customize/%';
+  $base_link['module'] = 'menu';
   $result = db_query("SELECT * FROM {menu_custom}");
   while ($menu = db_fetch_array($result)) {
+    // $link is passed by reference to menu_link_save(), so we make a copy of $base_link.
+    $link = $base_link;
     $link['mlid'] = 0;
     $link['link_title'] = $menu['title'];
     $link['link_path'] = 'admin/build/menu-customize/'. $menu['menu_name'];
@@ -296,7 +298,7 @@ function menu_nodeapi(&$node, $op) {
     case 'insert':
     case 'update':
       if (isset($node->menu)) {
-        $item = $node->menu;
+        $item = &$node->menu;
         if (!empty($item['delete'])) {
           menu_link_delete($item['mlid']);
         }
diff --git a/modules/node/node.info b/modules/node/node.info
index 3517b1a..dd63c9f 100644
--- a/modules/node/node.info
+++ b/modules/node/node.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/node/node.module b/modules/node/node.module
index 8ee80d0..468711d 100644
--- a/modules/node/node.module
+++ b/modules/node/node.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: node.module,v 1.947.2.13 2009/01/14 23:34:07 goba Exp $
+// $Id: node.module,v 1.947.2.15 2009/02/16 14:39:40 goba Exp $
 
 /**
  * @file
@@ -832,6 +832,7 @@ function node_submit($node) {
     }
     else {
       $node->teaser = '';
+      $node->format = 0;
     }
   }
 
@@ -1675,8 +1676,23 @@ function node_feed($nids = FALSE, $channel = array()) {
         $item = node_prepare($item, $teaser);
       }
 
-      // Allow modules to change $node->teaser before viewing.
+      // Allow modules to change $node->content before the node is rendered.
       node_invoke_nodeapi($item, 'view', $teaser, FALSE);
+
+      // Set the proper node property, then unset unused $node property so that a
+      // bad theme can not open a security hole.
+      $content = drupal_render($item->content);
+      if ($teaser) {
+        $item->teaser = $content;
+        unset($item->body);
+      }
+      else {
+        $item->body = $content;
+        unset($item->teaser);
+      }
+    
+      // Allow modules to modify the fully-built node.
+      node_invoke_nodeapi($item, 'alter', $teaser, FALSE);
     }
 
     // Allow modules to add additional item fields and/or modify $item
diff --git a/modules/openid/openid.info b/modules/openid/openid.info
index bece53e..96736f0 100644
--- a/modules/openid/openid.info
+++ b/modules/openid/openid.info
@@ -5,8 +5,8 @@ version = VERSION
 package = Core - optional
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/path/path.info b/modules/path/path.info
index 42a1711..e05299f 100644
--- a/modules/path/path.info
+++ b/modules/path/path.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/php/php.info b/modules/php/php.info
index 59b34dc..24873cf 100644
--- a/modules/php/php.info
+++ b/modules/php/php.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/ping/ping.info b/modules/ping/ping.info
index 603d20b..ba431eb 100644
--- a/modules/ping/ping.info
+++ b/modules/ping/ping.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/poll/poll.info b/modules/poll/poll.info
index fef456a..e757320 100644
--- a/modules/poll/poll.info
+++ b/modules/poll/poll.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/profile/profile.info b/modules/profile/profile.info
index ff79fca..d0a320b 100644
--- a/modules/profile/profile.info
+++ b/modules/profile/profile.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/search/search.info b/modules/search/search.info
index 47df322..cf170e4 100644
--- a/modules/search/search.info
+++ b/modules/search/search.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/search/search.module b/modules/search/search.module
index 22c9d0c..358899f 100644
--- a/modules/search/search.module
+++ b/modules/search/search.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: search.module,v 1.250.2.4 2008/09/17 06:42:20 goba Exp $
+// $Id: search.module,v 1.250.2.5 2009/02/25 16:17:40 goba Exp $
 
 /**
  * @file
@@ -637,7 +637,9 @@ function search_nodeapi(&$node, $op, $teaser = NULL, $page = NULL) {
       while ($link = db_fetch_object($result)) {
         $output[] = $link->caption;
       }
-      return '<a>('. implode(', ', $output) .')</a>';
+      if (count($output)) {
+        return '<a>('. implode(', ', $output) .')</a>';
+      }
     // Reindex the node when it is updated.  The node is automatically indexed
     // when it is added, simply by being added to the node table.
     case 'update':
diff --git a/modules/statistics/statistics.info b/modules/statistics/statistics.info
index bc945a2..d654b6c 100644
--- a/modules/statistics/statistics.info
+++ b/modules/statistics/statistics.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/syslog/syslog.info b/modules/syslog/syslog.info
index f7c67d2..0799538 100644
--- a/modules/syslog/syslog.info
+++ b/modules/syslog/syslog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/system/system.admin.inc b/modules/system/system.admin.inc
index ee8e350..8524efd 100644
--- a/modules/system/system.admin.inc
+++ b/modules/system/system.admin.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: system.admin.inc,v 1.63.2.6 2009/01/06 13:33:24 dries Exp $
+// $Id: system.admin.inc,v 1.63.2.7 2009/02/25 11:38:41 goba Exp $
 
 /**
  * @file
@@ -1726,7 +1726,7 @@ function system_run_cron() {
  * Menu callback: return information about PHP.
  */
 function system_php() {
-  phpinfo(INFO_GENERAL | INFO_CONFIGURATION);
+  phpinfo();
   exit();
 }
 
diff --git a/modules/system/system.info b/modules/system/system.info
index 86fc002..72b040f 100644
--- a/modules/system/system.info
+++ b/modules/system/system.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/system/system.install b/modules/system/system.install
index 2dcf397..8f689c9 100644
--- a/modules/system/system.install
+++ b/modules/system/system.install
@@ -1,5 +1,5 @@
 <?php
-// $Id: system.install,v 1.238.2.8 2009/01/14 21:36:16 goba Exp $
+// $Id: system.install,v 1.238.2.12 2009/02/25 14:02:46 goba Exp $
 
 /**
  * Test and report Drupal installation requirements.
@@ -64,10 +64,10 @@ function system_requirements($phase) {
   $memory_limit = ini_get('memory_limit');
   $requirements['php_memory_limit'] = array(
     'title' => $t('PHP memory limit'),
-    'value' => $memory_limit,
+    'value' => $memory_limit == -1 ? t('-1 (Unlimited)') : $memory_limit,
   );
 
-  if ($memory_limit && parse_size($memory_limit) < parse_size(DRUPAL_MINIMUM_PHP_MEMORY_LIMIT)) {
+  if ($memory_limit && $memory_limit != -1 && parse_size($memory_limit) < parse_size(DRUPAL_MINIMUM_PHP_MEMORY_LIMIT)) {
     $description = '';
     if ($phase == 'install') {
       $description = $t('Consider increasing your PHP memory limit to %memory_minimum_limit to help prevent errors in the installation process.', array('%memory_minimum_limit' => DRUPAL_MINIMUM_PHP_MEMORY_LIMIT));
@@ -1095,7 +1095,7 @@ function system_schema() {
       ),
     'unique keys' => array('dst_language' => array('dst', 'language')),
     'primary key' => array('pid'),
-    'indexes' => array('src' => array('src')),
+    'indexes' => array('src_language' => array('src', 'language')),
     );
 
   return $schema;
@@ -1161,7 +1161,7 @@ function system_update_6001() {
   db_add_primary_key($ret, 'term_node', array('vid', 'tid', 'nid'));
   db_add_index($ret, 'term_node', 'vid', array('vid'));
 
-  db_query('UPDATE {term_node} t SET vid = (SELECT vid FROM {node} n WHERE t.nid = n.nid)');
+  db_query('UPDATE {term_node} SET vid = (SELECT vid FROM {node} n WHERE {term_node}.nid = n.nid)');
   return $ret;
 }
 
@@ -1802,6 +1802,10 @@ function system_update_6021() {
         $item['router_path'] = $item['path'];
         $item['updated'] = FALSE;
       }
+      if ($item['description']) {
+        $item['options']['attributes']['title'] = $item['description'];
+      }      
+      
       // Save the link.
       menu_link_save($item);
       $_SESSION['menu_item_map'][$item['mid']] = array('mlid' => $item['mlid'], 'menu_name' => $item['menu_name']);
@@ -2534,6 +2538,16 @@ function system_update_6048() {
 
 
 /**
+ * Replace src index on the {url_alias} table with src, language.
+ */
+function system_update_6049() {
+  $ret = array();
+  db_drop_index($ret, 'url_alias', 'src');
+  db_add_index($ret, 'url_alias', 'src_language', array('src', 'language'));
+  return $ret;
+}
+
+/**
  * @} End of "defgroup updates-5.x-to-6.x"
  * The next series of updates should start at 7000.
  */
diff --git a/modules/system/system.module b/modules/system/system.module
index f5efd26..f58fdb2 100644
--- a/modules/system/system.module
+++ b/modules/system/system.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: system.module,v 1.585.2.30 2009/01/14 23:34:07 goba Exp $
+// $Id: system.module,v 1.585.2.32 2009/02/25 21:02:37 goba Exp $
 
 /**
  * @file
@@ -9,7 +9,7 @@
 /**
  * The current system version.
  */
-define('VERSION', '6.9');
+define('VERSION', '6.10');
 
 /**
  * Core API compatibility.
diff --git a/modules/taxonomy/taxonomy.admin.inc b/modules/taxonomy/taxonomy.admin.inc
index af425de..fa688f8 100644
--- a/modules/taxonomy/taxonomy.admin.inc
+++ b/modules/taxonomy/taxonomy.admin.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: taxonomy.admin.inc,v 1.22.2.2 2008/10/08 14:23:59 goba Exp $
+// $Id: taxonomy.admin.inc,v 1.22.2.3 2009/02/25 12:53:24 goba Exp $
 
 /**
  * @file
@@ -279,7 +279,7 @@ function taxonomy_overview_terms(&$form_state, $vocabulary) {
     // We are not calling taxonomy_get_tree because that might fail with a big
     // number of tags in the freetagging vocabulary.
     $results = pager_query(db_rewrite_sql('SELECT t.*, h.parent FROM {term_data} t INNER JOIN {term_hierarchy} h ON t.tid = h.tid WHERE t.vid = %d ORDER BY weight, name', 't', 'tid'), $page_increment, 0, NULL, $vocabulary->vid);
-    $total_entries = db_query(db_rewrite_sql('SELECT count(*) FROM {term_data} t INNER JOIN {term_hierarchy} h ON t.tid = h.tid WHERE t.vid = %d'), $page_increment, 0, NULL, $vocabulary->vid);
+    $total_entries = db_query(db_rewrite_sql('SELECT count(*) FROM {term_data} t INNER JOIN {term_hierarchy} h ON t.tid = h.tid WHERE t.vid = %d', 't', 'tid'), $page_increment, 0, NULL, $vocabulary->vid);
     while ($term = db_fetch_object($results)) {
       $key = 'tid:'. $term->tid .':0';
       $current_page[$key] = $term;
diff --git a/modules/taxonomy/taxonomy.info b/modules/taxonomy/taxonomy.info
index 99fb744..8d01b12 100644
--- a/modules/taxonomy/taxonomy.info
+++ b/modules/taxonomy/taxonomy.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/taxonomy/taxonomy.module b/modules/taxonomy/taxonomy.module
index 92a4492..09d3a20 100644
--- a/modules/taxonomy/taxonomy.module
+++ b/modules/taxonomy/taxonomy.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: taxonomy.module,v 1.414.2.5 2008/09/17 12:55:37 goba Exp $
+// $Id: taxonomy.module,v 1.414.2.6 2009/02/16 10:44:09 goba Exp $
 
 /**
  * @file
@@ -527,7 +527,7 @@ function taxonomy_form_alter(&$form, $form_state, $form_id) {
           '#default_value' => $typed_string,
           '#autocomplete_path' => 'taxonomy/autocomplete/'. $vocabulary->vid,
           '#weight' => $vocabulary->weight,
-          '#maxlength' => 255,
+          '#maxlength' => 1024,
         );
       }
       else {
diff --git a/modules/throttle/throttle.info b/modules/throttle/throttle.info
index 746784e..5879300 100644
--- a/modules/throttle/throttle.info
+++ b/modules/throttle/throttle.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/tracker/tracker.info b/modules/tracker/tracker.info
index 50f7a06..49db2f2 100644
--- a/modules/tracker/tracker.info
+++ b/modules/tracker/tracker.info
@@ -6,8 +6,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/translation/translation.info b/modules/translation/translation.info
index d079a84..86b9e9f 100644
--- a/modules/translation/translation.info
+++ b/modules/translation/translation.info
@@ -6,8 +6,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/trigger/trigger.info b/modules/trigger/trigger.info
index 93cd399..f9d172a 100644
--- a/modules/trigger/trigger.info
+++ b/modules/trigger/trigger.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/update/update.info b/modules/update/update.info
index 37c3027..b73bee8 100644
--- a/modules/update/update.info
+++ b/modules/update/update.info
@@ -5,8 +5,8 @@ version = VERSION
 package = Core - optional
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/upload/upload.info b/modules/upload/upload.info
index 1fab09e..48fb119 100644
--- a/modules/upload/upload.info
+++ b/modules/upload/upload.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/user/user.info b/modules/user/user.info
index 2204578..4044d91 100644
--- a/modules/user/user.info
+++ b/modules/user/user.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/modules/user/user.module b/modules/user/user.module
index b9bc071..5ada13d 100644
--- a/modules/user/user.module
+++ b/modules/user/user.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: user.module,v 1.892.2.11 2009/01/14 23:34:08 goba Exp $
+// $Id: user.module,v 1.892.2.12 2009/02/25 13:57:04 goba Exp $
 
 /**
  * @file
@@ -626,15 +626,15 @@ function user_user($type, &$edit, &$account, $category = NULL) {
   }
   if ($type == 'form' && $category == 'account') {
     $form_state = array();
-    return user_edit_form($form_state, arg(1), $edit);
+    return user_edit_form($form_state, (isset($account->uid) ? $account->uid : FALSE), $edit);
   }
 
   if ($type == 'validate' && $category == 'account') {
-    return _user_edit_validate(arg(1), $edit);
+    return _user_edit_validate((isset($account->uid) ? $account->uid : FALSE), $edit);
   }
 
   if ($type == 'submit' && $category == 'account') {
-    return _user_edit_submit(arg(1), $edit);
+    return _user_edit_submit((isset($account->uid) ? $account->uid : FALSE), $edit);
   }
 
   if ($type == 'categories') {
@@ -1441,7 +1441,8 @@ function user_edit_form(&$form_state, $uid, $edit, $register = FALSE) {
     '#title' => t('Account information'),
     '#weight' => -10,
   );
-  if (user_access('change own username') || $admin || $register) {
+  // Only show name field when: registration page; or user is editing own account and can change username; or an admin user.
+  if ($register || ($GLOBALS['user']->uid == $uid && user_access('change own username')) || $admin) {
     $form['account']['name'] = array('#type' => 'textfield',
       '#title' => t('Username'),
       '#default_value' => $edit['name'],
@@ -1543,9 +1544,8 @@ function user_edit_form(&$form_state, $uid, $edit, $register = FALSE) {
 }
 
 function _user_edit_validate($uid, &$edit) {
-  $user = user_load(array('uid' => $uid));
-  // Validate the username:
-  if (user_access('change own username') || user_access('administer users') || !$user->uid) {
+  // Validate the username when: new user account; or user is editing own account and can change username; or an admin user.
+  if (!$uid || ($GLOBALS['user']->uid == $uid && user_access('change own username')) || user_access('administer users')) {
     if ($error = user_validate_name($edit['name'])) {
       form_set_error('name', $error);
     }
@@ -1570,11 +1570,11 @@ function _user_edit_validate($uid, &$edit) {
 }
 
 function _user_edit_submit($uid, &$edit) {
-  $user = user_load(array('uid' => $uid));
+  $account = user_load($uid);
   // Delete picture if requested, and if no replacement picture was given.
   if (!empty($edit['picture_delete'])) {
-    if ($user->picture && file_exists($user->picture)) {
-      file_delete($user->picture);
+    if ($account->picture && file_exists($account->picture)) {
+      file_delete($account->picture);
     }
     $edit['picture'] = '';
   }
diff --git a/themes/bluemarine/bluemarine.info b/themes/bluemarine/bluemarine.info
index 1cee712..8a392bb 100644
--- a/themes/bluemarine/bluemarine.info
+++ b/themes/bluemarine/bluemarine.info
@@ -5,8 +5,8 @@ version = VERSION
 core = 6.x
 engine = phptemplate
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/themes/chameleon/chameleon.info b/themes/chameleon/chameleon.info
index ce33227..121bac7 100644
--- a/themes/chameleon/chameleon.info
+++ b/themes/chameleon/chameleon.info
@@ -12,8 +12,8 @@ stylesheets[all][] = common.css
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/themes/chameleon/marvin/marvin.info b/themes/chameleon/marvin/marvin.info
index 2079179..30c46ef 100644
--- a/themes/chameleon/marvin/marvin.info
+++ b/themes/chameleon/marvin/marvin.info
@@ -7,8 +7,8 @@ version = VERSION
 core = 6.x
 base theme = chameleon
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/themes/garland/garland.info b/themes/garland/garland.info
index 21b907a..d23b8e6 100644
--- a/themes/garland/garland.info
+++ b/themes/garland/garland.info
@@ -7,8 +7,8 @@ engine = phptemplate
 stylesheets[all][] = style.css
 stylesheets[print][] = print.css
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/themes/garland/minnelli/minnelli.info b/themes/garland/minnelli/minnelli.info
index ebf275f..59d4afe 100644
--- a/themes/garland/minnelli/minnelli.info
+++ b/themes/garland/minnelli/minnelli.info
@@ -6,8 +6,8 @@ core = 6.x
 base theme = garland
 stylesheets[all][] = minnelli.css
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
diff --git a/themes/garland/template.php b/themes/garland/template.php
index 1c7faf2..c3f3b43 100644
--- a/themes/garland/template.php
+++ b/themes/garland/template.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: template.php,v 1.16 2007/10/11 09:51:29 goba Exp $
+// $Id: template.php,v 1.16.2.1 2009/02/25 11:47:37 goba Exp $
 
 /**
  * Sets the body-tag class attribute.
@@ -94,7 +94,7 @@ function phptemplate_get_ie_styles() {
   global $language;
 
   $iecss = '<link type="text/css" rel="stylesheet" media="all" href="'. base_path() . path_to_theme() .'/fix-ie.css" />';
-  if (defined('LANGUAGE_RTL') && $language->direction == LANGUAGE_RTL) {
+  if ($language->direction == LANGUAGE_RTL) {
     $iecss .= '<style type="text/css" media="all">@import "'. base_path() . path_to_theme() .'/fix-ie-rtl.css";</style>';
   }
 
diff --git a/themes/pushbutton/pushbutton.info b/themes/pushbutton/pushbutton.info
index e237af6..c82ca7a 100644
--- a/themes/pushbutton/pushbutton.info
+++ b/themes/pushbutton/pushbutton.info
@@ -5,8 +5,8 @@ version = VERSION
 core = 6.x
 engine = phptemplate
 
-; Information added by drupal.org packaging script on 2009-01-14
-version = "6.9"
+; Information added by drupal.org packaging script on 2009-02-25
+version = "6.10"
 project = "drupal"
-datestamp = "1231976413"
+datestamp = "1235596218"
 
