diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index 69d02fc..9615555 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,4 +1,12 @@
-// $Id: CHANGELOG.txt,v 1.253.2.25 2009/02/25 21:02:36 goba Exp $
+// $Id: CHANGELOG.txt,v 1.253.2.27 2009/04/30 00:13:30 goba Exp $
+
+Drupal 6.11, 2009-04-29
+----------------------
+- Fixed security issues (Cross site scripting and limited information
+  disclosure), see SA-CORE-2009-005
+- Fixed performance issues with the menu router cache, the update
+  status cache and improved cache invalidation
+- Fixed a variety of small bugs.
 
 Drupal 6.10, 2009-02-25
 ----------------------
@@ -173,6 +181,16 @@ Drupal 6.0, 2008-02-13
 - Removed old system updates. Updates from Drupal versions prior to 5.x will
   require upgrading to 5.x before upgrading to 6.x.
 
+Drupal 5.17, 2009-04-29
+-----------------------
+- Fixed security issues (Cross site scripting and limited information disclosure) see SA-CORE-2009-005.
+- Fixed a variety of small bugs.
+
+Drupal 5.16, 2009-02-25
+-----------------------
+- Fixed a security issue, (Local file inclusion on Windows), see SA-CORE-2009-004.
+- Fixed a variety of small bugs.
+
 Drupal 5.15, 2009-01-14
 ----------------------
 - Fixed security issues, (Hardening against SQL injection), see SA-CORE-2009-001
diff --git a/MAINTAINERS.txt b/MAINTAINERS.txt
index 76e783b..7a3a200 100644
--- a/MAINTAINERS.txt
+++ b/MAINTAINERS.txt
@@ -1,4 +1,4 @@
-// $Id: MAINTAINERS.txt,v 1.19.2.1 2008/05/15 22:13:42 dries Exp $
+// $Id: MAINTAINERS.txt,v 1.19.2.2 2009/04/29 17:15:10 goba Exp $
 
 List of maintainers
 --------------------------------------------------------------------------------
@@ -72,7 +72,6 @@ S: maintained
 
 UPDATE MODULE
 M: Derek Wright <http://drupal.org/user/46549/contact>
-   Earl Miles <http://drupal.org/user/26979/contact>
 S: maintained
 
 XML-RPC SERVER/CLIENT
diff --git a/includes/bootstrap.inc b/includes/bootstrap.inc
index 70e5156..1d1caad 100644
--- a/includes/bootstrap.inc
+++ b/includes/bootstrap.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: bootstrap.inc,v 1.206.2.11 2009/02/25 13:49:54 dries Exp $
+// $Id: bootstrap.inc,v 1.206.2.12 2009/04/30 00:13:30 goba Exp $
 
 /**
  * @file
@@ -791,6 +791,8 @@ function request_uri() {
       $uri = $_SERVER['SCRIPT_NAME'];
     }
   }
+  // Prevent multiple slashes to avoid cross site requests via the FAPI.
+  $uri = '/'. ltrim($uri, '/');
 
   return $uri;
 }
diff --git a/includes/cache.inc b/includes/cache.inc
index 054d840..50faf6d 100644
--- a/includes/cache.inc
+++ b/includes/cache.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: cache.inc,v 1.17 2008/01/29 11:36:06 goba Exp $
+// $Id: cache.inc,v 1.17.2.1 2009/04/27 14:23:58 goba Exp $
 
 /**
  * Return data from the persistent cache. Data may be stored as either plain text or as serialized data.
@@ -15,10 +15,10 @@ function cache_get($cid, $table = 'cache') {
   global $user;
 
   // Garbage collection necessary when enforcing a minimum cache lifetime
-  $cache_flush = variable_get('cache_flush', 0);
+  $cache_flush = variable_get('cache_flush_'. $table, 0);
   if ($cache_flush && ($cache_flush + variable_get('cache_lifetime', 0) <= time())) {
     // Reset the variable immediately to prevent a meltdown in heavy load situations.
-    variable_set('cache_flush', 0);
+    variable_set('cache_flush_'. $table, 0);
     // Time to flush old cache data
     db_query("DELETE FROM {". $table ."} WHERE expire != %d AND expire <= %d", CACHE_PERMANENT, $cache_flush);
   }
@@ -149,16 +149,16 @@ function cache_clear_all($cid = NULL, $table = NULL, $wildcard = FALSE) {
       // cached data that was cached before the timestamp.
       $user->cache = time();
 
-      $cache_flush = variable_get('cache_flush', 0);
+      $cache_flush = variable_get('cache_flush_'. $table, 0);
       if ($cache_flush == 0) {
         // This is the first request to clear the cache, start a timer.
-        variable_set('cache_flush', time());
+        variable_set('cache_flush_'. $table, time());
       }
       else if (time() > ($cache_flush + variable_get('cache_lifetime', 0))) {
         // Clear the cache for everyone, cache_flush_delay seconds have
         // passed since the first request to clear the cache.
         db_query("DELETE FROM {". $table ."} WHERE expire != %d AND expire < %d", CACHE_PERMANENT, time());
-        variable_set('cache_flush', 0);
+        variable_set('cache_flush_'. $table, 0);
       }
     }
     else {
diff --git a/includes/common.inc b/includes/common.inc
index 8d240db..c4a19ad 100644
--- a/includes/common.inc
+++ b/includes/common.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: common.inc,v 1.756.2.47 2009/02/25 21:02:36 goba Exp $
+// $Id: common.inc,v 1.756.2.49 2009/04/30 00:13:30 goba Exp $
 
 /**
  * @file
@@ -152,6 +152,15 @@ function drupal_get_headers() {
 }
 
 /**
+ * Make any final alterations to the rendered xhtml.
+ */
+function drupal_final_markup($content) {
+  // Make sure that the charset is always specified as the first element of the
+  // head region to prevent encoding-based attacks.
+  return preg_replace('/<head[^>]*>/i', "\$0\n<meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\" />", $content, 1);
+}
+
+/**
  * Add a feed URL for the current page.
  *
  * @param $url
diff --git a/includes/database.pgsql.inc b/includes/database.pgsql.inc
index bc267e8..93b2d31 100644
--- a/includes/database.pgsql.inc
+++ b/includes/database.pgsql.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: database.pgsql.inc,v 1.68.2.3 2008/11/17 10:31:41 goba Exp $
+// $Id: database.pgsql.inc,v 1.68.2.4 2009/03/30 13:04:06 goba Exp $
 
 /**
  * @file
@@ -474,7 +474,7 @@ function db_type_map() {
     'blob:big' => 'bytea',
     'blob:normal' => 'bytea',
 
-    'datetime:normal' => 'timestamp',
+    'datetime:normal' => 'timestamp without time zone',
 
     'serial:tiny' => 'serial',
     'serial:small' => 'serial',
diff --git a/includes/file.inc b/includes/file.inc
index 7e5b697..78536c3 100644
--- a/includes/file.inc
+++ b/includes/file.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: file.inc,v 1.121.2.5 2008/10/20 09:42:31 goba Exp $
+// $Id: file.inc,v 1.121.2.6 2009/04/13 19:07:16 dries Exp $
 
 /**
  * @file
@@ -463,8 +463,8 @@ function file_space_used($uid = NULL) {
  * @param $validators
  *   An optional, associative array of callback functions used to validate the
  *   file. The keys are function names and the values arrays of callback
- *   parameters which will be passed in after the user and file objects. The
- *   functions should return an array of error messages, an empty array
+ *   parameters which will be passed in after the file object. The
+ *   functions should return an array of error messages; an empty array
  *   indicates that the file passed validation. The functions will be called in
  *   the order specified.
  * @param $dest
diff --git a/includes/form.inc b/includes/form.inc
index fb1691e..c3056f0 100644
--- a/includes/form.inc
+++ b/includes/form.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: form.inc,v 1.265.2.19 2009/02/22 18:12:46 goba Exp $
+// $Id: form.inc,v 1.265.2.23 2009/04/27 14:41:11 goba Exp $
 
 /**
  * @defgroup forms Form builder functions
@@ -624,6 +624,12 @@ function drupal_redirect_form($form, $redirect = NULL) {
   }
   if (!isset($goto) || ($goto !== FALSE)) {
     if (isset($goto)) {
+      // Remove any fake destination set by drupal_not_found() or
+      // drupal_access_denied() so that we can properly redirect from those
+      // pages.
+      if (isset($_REQUEST['destination']) && $_REQUEST['destination'] == $_GET['q']) {
+        unset($_REQUEST['destination']);
+      }
       if (is_array($goto)) {
         call_user_func_array('drupal_goto', $goto);
       }
@@ -756,7 +762,10 @@ function form_execute_handlers($type, &$form, &$form_state) {
 
   foreach ($handlers as $function) {
     if (function_exists($function))  {
-      if ($type == 'submit' && ($batch =& batch_get())) {
+      // Check to see if a previous _submit handler has set a batch, but 
+      // make sure we do not react to a batch that is already being processed 
+      // (for instance if a batch operation performs a drupal_execute()).
+      if ($type == 'submit' && ($batch =& batch_get()) && !isset($batch['current_set'])) {
         // Some previous _submit handler has set a batch. We store the call
         // in a special 'control' batch set, for execution at the correct
         // time during the batch processing workflow.
diff --git a/includes/locale.inc b/includes/locale.inc
index 4ce2d78..64fc8d0 100644
--- a/includes/locale.inc
+++ b/includes/locale.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: locale.inc,v 1.174.2.7 2009/01/06 15:36:51 goba Exp $
+// $Id: locale.inc,v 1.174.2.8 2009/04/08 02:54:00 dries Exp $
 
 /**
  * @file
@@ -2433,7 +2433,7 @@ function _locale_get_predefined_list() {
     "se" => array("Northern Sami"),
     "sg" => array("Sango"),
     "sh" => array("Serbo-Croatian"),
-    "si" => array("Singhalese"),
+    "si" => array("Sinhala", "සිංහල"),
     "sk" => array("Slovak", "Slovenčina"),
     "sl" => array("Slovenian", "Slovenščina"),
     "sm" => array("Samoan"),
diff --git a/includes/mail.inc b/includes/mail.inc
index bfbdd82..830f89b 100644
--- a/includes/mail.inc
+++ b/includes/mail.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: mail.inc,v 1.8.2.4 2008/10/06 11:04:08 dries Exp $
+// $Id: mail.inc,v 1.8.2.6 2009/04/27 11:07:43 goba Exp $
 
 /**
  * Compose and optionally send an e-mail message.
@@ -70,7 +70,7 @@
  * @param $params
  *   Optional parameters to build the e-mail.
  * @param $from
- *   Sets From, Reply-To, Return-Path and Error-To to this value, if given.
+ *   Sets From to this value, if given.
  * @param $send
  *   Send the message directly, without calling drupal_mail_send() manually.
  * @return
@@ -105,10 +105,10 @@ function drupal_mail($module, $key, $to, $language, $params = array(), $from = N
     // To prevent e-mail from looking like spam, the addresses in the Sender and
     // Return-Path headers should have a domain authorized to use the originating
     // SMTP server. Errors-To is redundant, but shouldn't hurt.
-    $headers['From'] = $headers['Reply-To'] = $headers['Sender'] = $headers['Return-Path'] = $headers['Errors-To'] = $default_from;
+    $headers['From'] = $headers['Sender'] = $headers['Return-Path'] = $headers['Errors-To'] = $default_from;
   }
   if ($from) {
-    $headers['From'] = $headers['Reply-To'] = $from;
+    $headers['From'] = $from;
   }
   $message['headers'] = $headers;
 
@@ -132,7 +132,7 @@ function drupal_mail($module, $key, $to, $language, $params = array(), $from = N
     // Log errors
     if (!$message['result']) {
       watchdog('mail', 'Error sending e-mail (from %from to %to).', array('%from' => $message['from'], '%to' => $message['to']), WATCHDOG_ERROR);
-      drupal_set_message(t('Unable to send e-mail. Please contact the site admin, if the problem persists.'), 'error');
+      drupal_set_message(t('Unable to send e-mail. Please contact the site administrator if the problem persists.'), 'error');
     }
   }
 
diff --git a/includes/menu.inc b/includes/menu.inc
index 29635e6..7ab59ba 100644
--- a/includes/menu.inc
+++ b/includes/menu.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: menu.inc,v 1.255.2.28 2009/02/09 16:28:21 dries Exp $
+// $Id: menu.inc,v 1.255.2.31 2009/04/27 12:50:13 goba Exp $
 
 /**
  * @file
@@ -482,9 +482,13 @@ function _menu_check_access(&$item, $map) {
 function _menu_item_localize(&$item, $map, $link_translate = FALSE) {
   $callback = $item['title_callback'];
   $item['localized_options'] = $item['options'];
-  // If we are not doing link translation or if the title matches the
-  // link title of its router item, localize it.
-  if (!$link_translate || (!empty($item['title']) && ($item['title'] == $item['link_title']))) {
+  // If we are translating the title of a menu link, and its title is the same
+  // as the corresponding router item, then we can use the title information
+  // from the router. If it's customized, then we need to use the link title
+  // itself; can't localize.
+  // If we are translating a router item (tabs, page, breadcrumb), then we
+  // can always use the information from the router item.
+  if (!$link_translate || ($item['title'] == $item['link_title'])) {
     // t() is a special case. Since it is used very close to all the time,
     // we handle it directly instead of using indirect, slower methods.
     if ($callback == 't') {
@@ -963,6 +967,11 @@ function _menu_tree_cid($menu_name, $data) {
 
 /**
  * Recursive helper function - collect node links.
+ *
+ * @param $tree
+ *   The menu tree you wish to collect node links from.
+ * @param $node_links
+ *   An array in which to store the collected node links.
  */
 function menu_tree_collect_node_links(&$tree, &$node_links) {
   foreach ($tree as $key => $v) {
@@ -981,6 +990,12 @@ function menu_tree_collect_node_links(&$tree, &$node_links) {
 
 /**
  * Check access and perform other dynamic operations for each link in the tree.
+ *
+ * @param $tree
+ *   The menu tree you wish to operate on.
+ * @param $node_links
+ *   A collection of node link references generated from $tree by
+ *   menu_tree_collect_node_links().
  */
 function menu_tree_check_access(&$tree, $node_links = array()) {
 
@@ -1655,10 +1670,10 @@ function menu_cache_clear_all() {
  */
 function menu_rebuild() {
   variable_del('menu_rebuild_needed');
-  menu_cache_clear_all();
   $menu = menu_router_build(TRUE);
   _menu_navigation_links_rebuild($menu);
-  // Clear the page and block caches.
+  // Clear the menu, page and block caches.
+  menu_cache_clear_all();
   _menu_clear_page_cache();
   if (defined('MAINTENANCE_MODE')) {
     variable_set('menu_rebuild_needed', TRUE);
@@ -1672,26 +1687,34 @@ function menu_router_build($reset = FALSE) {
   static $menu;
 
   if (!isset($menu) || $reset) {
-    if (!$reset && ($cache = cache_get('router:', 'cache_menu')) && isset($cache->data)) {
-      $menu = $cache->data;
-    }
-    else {
-      // We need to manually call each module so that we can know which module
-      // a given item came from.
-      $callbacks = array();
-      foreach (module_implements('menu') as $module) {
-        $router_items = call_user_func($module .'_menu');
-        if (isset($router_items) && is_array($router_items)) {
-          foreach (array_keys($router_items) as $path) {
-            $router_items[$path]['module'] = $module;
-          }
-          $callbacks = array_merge($callbacks, $router_items);
+    // We need to manually call each module so that we can know which module
+    // a given item came from.
+    $callbacks = array();
+    foreach (module_implements('menu') as $module) {
+      $router_items = call_user_func($module .'_menu');
+      if (isset($router_items) && is_array($router_items)) {
+        foreach (array_keys($router_items) as $path) {
+          $router_items[$path]['module'] = $module;
         }
+        $callbacks = array_merge($callbacks, $router_items);
       }
-      // Alter the menu as defined in modules, keys are like user/%user.
-      drupal_alter('menu', $callbacks);
-      $menu = _menu_router_build($callbacks);
     }
+    // Alter the menu as defined in modules, keys are like user/%user.
+    drupal_alter('menu', $callbacks);
+    $menu = _menu_router_build($callbacks);
+    _menu_router_cache($menu);
+  }
+  return $menu;
+}
+
+/**
+ * Helper function to store the menu router if we have it in memory.
+ */
+function _menu_router_cache($new_menu = NULL) {
+  static $menu = NULL;
+
+  if (isset($new_menu)) {
+    $menu = $new_menu;
   }
   return $menu;
 }
@@ -1758,7 +1781,7 @@ function _menu_navigation_links_rebuild($menu) {
   // Updated and customized items whose router paths are gone need new ones.
   $result = db_query("SELECT ml.link_path, ml.mlid, ml.router_path, ml.updated FROM {menu_links} ml WHERE ml.updated = 1 OR (router_path NOT IN ($placeholders) AND external = 0 AND customized = 1)", $paths);
   while ($item = db_fetch_array($result)) {
-    $router_path = _menu_find_router_path($menu, $item['link_path']);
+    $router_path = _menu_find_router_path($item['link_path']);
     if (!empty($router_path) && ($router_path != $item['router_path'] || $item['updated'])) {
       // If the router path and the link path matches, it's surely a working
       // item, so we clear the updated flag.
@@ -1841,8 +1864,10 @@ function _menu_delete_item($item, $force = FALSE) {
  *   saved.
  */
 function menu_link_save(&$item) {
-  $menu = menu_router_build();
 
+  // Get the router if it's already in memory. $menu will be NULL, unless this
+  // is during a menu rebuild
+  $menu = _menu_router_cache();
   drupal_alter('menu_link', $item, $menu);
 
   // This is the easiest way to handle the unique internal path '<front>',
@@ -1960,7 +1985,7 @@ function menu_link_save(&$item) {
     else {
       // Find the router path which will serve this path.
       $item['parts'] = explode('/', $item['link_path'], MENU_MAX_PARTS);
-      $item['router_path'] = _menu_find_router_path($menu, $item['link_path']);
+      $item['router_path'] = _menu_find_router_path($item['link_path']);
     }
   }
   db_query("UPDATE {menu_links} SET menu_name = '%s', plid = %d, link_path = '%s',
@@ -2021,25 +2046,35 @@ function _menu_set_expanded_menus() {
 /**
  * Find the router path which will serve this path.
  *
- * @param $menu
- *  The full built menu.
  * @param $link_path
  *  The path for we are looking up its router path.
  * @return
  *  A path from $menu keys or empty if $link_path points to a nonexisting
  *  place.
  */
-function _menu_find_router_path($menu, $link_path) {
-  $parts = explode('/', $link_path, MENU_MAX_PARTS);
+function _menu_find_router_path($link_path) {
+  // $menu will only have data during a menu rebuild.
+  $menu = _menu_router_cache();
+
   $router_path = $link_path;
-  if (!isset($menu[$router_path])) {
-    list($ancestors) = menu_get_ancestors($parts);
+  $parts = explode('/', $link_path, MENU_MAX_PARTS);
+  list($ancestors, $placeholders) = menu_get_ancestors($parts);
+
+  if (empty($menu)) {
+    // Not during a menu rebuild, so look up in the database.
+    $router_path = (string)db_result(db_query_range('SELECT path FROM {menu_router} WHERE path IN ('. implode (',', $placeholders) .') ORDER BY fit DESC', $ancestors, 0, 1));
+  }
+  elseif (!isset($menu[$router_path])) {
+    // Add an empty path as a fallback.
     $ancestors[] = '';
     foreach ($ancestors as $key => $router_path) {
       if (isset($menu[$router_path])) {
+        // Exit the loop leaving $router_path as the first match.
         break;
       }
     }
+    // If we did not find the path, $router_path will be the empty string
+    // at the end of $ancestors.
   }
   return $router_path;
 }
@@ -2391,7 +2426,7 @@ function _menu_router_build($callbacks) {
   $masks = array_keys($masks);
   rsort($masks);
   variable_set('menu_masks', $masks);
-  cache_set('router:', $menu, 'cache_menu');
+
   return $menu;
 }
 
diff --git a/includes/theme.inc b/includes/theme.inc
index 1abe305..4b92e53 100644
--- a/includes/theme.inc
+++ b/includes/theme.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: theme.inc,v 1.415.2.19 2009/02/25 21:02:16 goba Exp $
+// $Id: theme.inc,v 1.415.2.21 2009/04/30 00:13:30 goba Exp $
 
 /**
  * @file
@@ -687,6 +687,10 @@ function theme() {
   }
   // restore path_to_theme()
   $theme_path = $temp;
+  // Add final markup to the full page.
+  if ($hook == 'page') {
+    $output = drupal_final_markup($output);
+  }
   return $output;
 }
 
@@ -698,11 +702,16 @@ function theme() {
 function drupal_discover_template($paths, $suggestions, $extension = '.tpl.php') {
   global $theme_engine;
 
+  // Remove slashes or null to prevent files from being included from
+  // an unexpected location (especially on Windows servers).
+  $extension = str_replace(array("/", "\\", "\0"), '', $extension);
+
   // Loop through all paths and suggestions in FIFO order.
   $suggestions = array_reverse($suggestions);
   $paths = array_reverse($paths);
   foreach ($suggestions as $suggestion) {
     if (!empty($suggestion)) {
+      $suggestion = str_replace(array("/", "\\", "\0"), '', $suggestion);
       foreach ($paths as $path) {
         if (file_exists($file = $path .'/'. $suggestion . $extension)) {
           return $file;
@@ -1868,7 +1877,7 @@ function template_preprocess_page(&$variables) {
   $suggestion = 'page';
   $suggestions = array();
   while ($arg = arg($i++)) {
-    $arg = str_replace(array('/', '\\', '\0'), '', $arg);
+    $arg = str_replace(array("/", "\\", "\0"), '', $arg);
     $suggestions[] = $suggestion .'-'. $arg;
     if (!is_numeric($arg)) {
       $suggestion .= '-'. $arg;
diff --git a/install.php b/install.php
index a13f8b8..e0d734d 100644
--- a/install.php
+++ b/install.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: install.php,v 1.113.2.8 2009/02/25 11:47:36 goba Exp $
+// $Id: install.php,v 1.113.2.9 2009/04/27 10:50:35 goba Exp $
 
 require_once './includes/install.inc';
 
@@ -257,7 +257,6 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Database name'),
       '#default_value' => $db_path,
       '#size' => 45,
-      '#maxlength' => 45,
       '#required' => TRUE,
       '#description' => $db_path_description
     );
@@ -268,7 +267,6 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Database username'),
       '#default_value' => $db_user,
       '#size' => 45,
-      '#maxlength' => 45,
       '#required' => TRUE,
     );
 
@@ -278,7 +276,6 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Database password'),
       '#default_value' => $db_pass,
       '#size' => 45,
-      '#maxlength' => 45,
     );
 
     $form['advanced_options'] = array(
@@ -295,7 +292,8 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Database host'),
       '#default_value' => $db_host,
       '#size' => 45,
-      '#maxlength' => 45,
+      // Hostnames can be 255 characters long.
+      '#maxlength' => 255,
       '#required' => TRUE,
       '#description' => st('If your database is located on a different server, change this.'),
     );
@@ -306,7 +304,8 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Database port'),
       '#default_value' => $db_port,
       '#size' => 45,
-      '#maxlength' => 45,
+      // The maximum port number is 65536, 5 digits.
+      '#maxlength' => 5,
       '#description' => st('If your database server is listening to a non-standard port, enter its number.'),
     );
 
@@ -317,7 +316,6 @@ function install_settings_form(&$form_state, $profile, $install_locale, $setting
       '#title' => st('Table prefix'),
       '#default_value' => $db_prefix,
       '#size' => 45,
-      '#maxlength' => 45,
       '#description' => st('If more than one application will be sharing this database, enter a table prefix such as %prefix for your @drupal site here.', array('@drupal' => drupal_install_profile_name(), '%prefix' => $prefix)),
     );
 
diff --git a/misc/tableheader.js b/misc/tableheader.js
index 037d815..f572e29 100644
--- a/misc/tableheader.js
+++ b/misc/tableheader.js
@@ -1,4 +1,4 @@
-// $Id: tableheader.js,v 1.16.2.1 2008/10/01 23:30:36 goba Exp $
+// $Id: tableheader.js,v 1.16.2.2 2009/03/30 12:48:09 goba Exp $
 
 Drupal.tableHeaderDoScroll = function() {
   if (typeof(Drupal.tableHeaderOnScroll)=='function') {
@@ -70,8 +70,12 @@ Drupal.behaviors.tableHeader = function (context) {
     // Get the height of the header table and scroll up that amount.
     if (prevAnchor != location.hash) {
       if (location.hash != '') {
-        var scrollLocation = $('td'+ location.hash).offset().top - $(e).height();
-        $('body, html').scrollTop(scrollLocation);
+        var offset = $('td' + location.hash).offset();
+        if (offset) {
+          var top = offset.top;
+          var scrollLocation = top - $(e).height();
+          $('body, html').scrollTop(scrollLocation);
+        }
       }
       prevAnchor = location.hash;
     }
diff --git a/modules/aggregator/aggregator.info b/modules/aggregator/aggregator.info
index 053055d..ba7ec3b 100644
--- a/modules/aggregator/aggregator.info
+++ b/modules/aggregator/aggregator.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/aggregator/aggregator.module b/modules/aggregator/aggregator.module
index a2e2aa4..858fd13 100644
--- a/modules/aggregator/aggregator.module
+++ b/modules/aggregator/aggregator.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: aggregator.module,v 1.374.2.4 2009/01/14 21:36:16 goba Exp $
+// $Id: aggregator.module,v 1.374.2.5 2009/03/30 12:15:53 goba Exp $
 
 /**
  * @file
@@ -390,7 +390,7 @@ function aggregator_save_category($edit) {
   else if (!empty($edit['title'])) {
     // A single unique id for bundles and feeds, to use in blocks
     db_query("INSERT INTO {aggregator_category} (title, description, block) VALUES ('%s', '%s', 5)", $edit['title'], $edit['description']);
-    $link_path .= db_last_insert_id('aggregator', 'cid');
+    $link_path .= db_last_insert_id('aggregator_category', 'cid');
     $op = 'insert';
   }
   if (isset($op)) {
diff --git a/modules/block/block.info b/modules/block/block.info
index 527c728..ffa9e7e 100644
--- a/modules/block/block.info
+++ b/modules/block/block.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/blog/blog.info b/modules/blog/blog.info
index 86c8eb3..4867a67 100644
--- a/modules/blog/blog.info
+++ b/modules/blog/blog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/blogapi/blogapi.info b/modules/blogapi/blogapi.info
index b62897d..8870507 100644
--- a/modules/blogapi/blogapi.info
+++ b/modules/blogapi/blogapi.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/book/book.info b/modules/book/book.info
index e6e7815..0b7f18d 100644
--- a/modules/book/book.info
+++ b/modules/book/book.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/color/color.info b/modules/color/color.info
index 6b15d1b..d7a9c0f 100644
--- a/modules/color/color.info
+++ b/modules/color/color.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/comment/comment.info b/modules/comment/comment.info
index 6ff2108..fe01989 100644
--- a/modules/comment/comment.info
+++ b/modules/comment/comment.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/comment/comment.module b/modules/comment/comment.module
index 1505523..ebd6be7 100644
--- a/modules/comment/comment.module
+++ b/modules/comment/comment.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: comment.module,v 1.617.2.4 2009/01/06 17:34:54 goba Exp $
+// $Id: comment.module,v 1.617.2.6 2009/04/05 12:04:52 dries Exp $
 
 /**
  * @file
@@ -1051,7 +1051,9 @@ function comment_render($node, $cid = 0) {
       $output .= comment_form_box(array('nid' => $nid), t('Post new comment'));
     }
 
-    $output = theme('comment_wrapper', $output, $node);
+    if ($output) {
+      $output = theme('comment_wrapper', $output, $node);
+    }
   }
 
   return $output;
@@ -1523,7 +1525,7 @@ function _comment_form_submit(&$comment_values) {
     // 2) Strip out all HTML tags
     // 3) Convert entities back to plain-text.
     // Note: format is checked by check_markup().
-    $comment_values['subject'] = trim(truncate_utf8(decode_entities(strip_tags(check_markup($comment_values['comment'], $comment_values['format']))), 29, TRUE));
+    $comment_values['subject'] = truncate_utf8(trim(decode_entities(strip_tags(check_markup($comment_values['comment'], $comment_values['format'])))), 29, TRUE);
     // Edge cases where the comment body is populated only by HTML tags will
     // require a default subject.
     if ($comment_values['subject'] == '') {
diff --git a/modules/contact/contact.info b/modules/contact/contact.info
index 1b7f2da..7257692 100644
--- a/modules/contact/contact.info
+++ b/modules/contact/contact.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/contact/contact.pages.inc b/modules/contact/contact.pages.inc
index 79973b1..7c6b8b9 100644
--- a/modules/contact/contact.pages.inc
+++ b/modules/contact/contact.pages.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: contact.pages.inc,v 1.6.2.1 2008/02/12 14:42:50 goba Exp $
+// $Id: contact.pages.inc,v 1.6.2.2 2009/04/27 14:48:21 goba Exp $
 
 /**
  * @file
@@ -177,11 +177,11 @@ function contact_mail_user(&$form_state, $recipient) {
   $form['recipient'] = array('#type' => 'value', '#value' => $recipient);
   $form['from'] = array('#type' => 'item',
     '#title' => t('From'),
-    '#value' => check_plain($user->name) .' &lt;'. check_plain($user->mail) .'&gt;',
+    '#value' => theme('username', $user) .' &lt;'. check_plain($user->mail) .'&gt;',
   );
   $form['to'] = array('#type' => 'item',
     '#title' => t('To'),
-    '#value' => check_plain($recipient->name),
+    '#value' => theme('username', $recipient),
   );
   $form['subject'] = array('#type' => 'textfield',
     '#title' => t('Subject'),
diff --git a/modules/dblog/dblog.info b/modules/dblog/dblog.info
index 5c31697..b59f108 100644
--- a/modules/dblog/dblog.info
+++ b/modules/dblog/dblog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/filter/filter.info b/modules/filter/filter.info
index 19b6987..e8e4854 100644
--- a/modules/filter/filter.info
+++ b/modules/filter/filter.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/forum/forum.info b/modules/forum/forum.info
index c450515..c36d76e 100644
--- a/modules/forum/forum.info
+++ b/modules/forum/forum.info
@@ -7,8 +7,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/forum/forum.module b/modules/forum/forum.module
index ff28bb3..fd650aa 100644
--- a/modules/forum/forum.module
+++ b/modules/forum/forum.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: forum.module,v 1.448.2.5 2008/10/24 19:07:47 dries Exp $
+// $Id: forum.module,v 1.448.2.6 2009/03/30 11:09:51 goba Exp $
 
 /**
  * @file
@@ -468,19 +468,11 @@ function forum_form(&$node, $form_state) {
   return $form;
 }
 
-function forum_link_alter(&$links, $node) {
-  foreach ($links as $module => $link) {
-    if (strstr($module, 'taxonomy_term')) {
-      // Link back to the forum and not the taxonomy term page. We'll only
-      // do this if the taxonomy term in question belongs to forums.
-      $tid = str_replace('taxonomy/term/', '', $link['href']);
-      $vid = variable_get('forum_nav_vocabulary', '');
-      $term = taxonomy_get_term($tid);
-      if ($term->vid == $vid) {
-        $links[$module]['href'] = str_replace('taxonomy/term', 'forum', $link['href']);
-      }
-    }
-  }
+/**
+ * Implementation of hook_term_path().
+ */
+function forum_term_path($term) {
+  return 'forum/'. $term->tid;
 }
 
 /**
diff --git a/modules/help/help.info b/modules/help/help.info
index 1d237b4..0163b54 100644
--- a/modules/help/help.info
+++ b/modules/help/help.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/locale/locale.info b/modules/locale/locale.info
index 0efb98e..fa165d8 100644
--- a/modules/locale/locale.info
+++ b/modules/locale/locale.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/menu/menu.info b/modules/menu/menu.info
index 4060db1..08446da 100644
--- a/modules/menu/menu.info
+++ b/modules/menu/menu.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/menu/menu.module b/modules/menu/menu.module
index 497abbd..b25d173 100644
--- a/modules/menu/menu.module
+++ b/modules/menu/menu.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: menu.module,v 1.157.2.5 2009/02/25 13:15:40 goba Exp $
+// $Id: menu.module,v 1.157.2.6 2009/04/27 12:50:13 goba Exp $
 
 /**
  * @file
@@ -256,8 +256,7 @@ function _menu_parents_recurse($tree, $menu_name, $indent, &$options, $exclude,
  * Reset a system-defined menu item.
  */
 function menu_reset_item($item) {
-  $router = menu_router_build();
-  $new_item = _menu_link_build($router[$item['router_path']]);
+  $new_item = _menu_link_build(menu_get_item($item['router_path']));
   foreach (array('mlid', 'has_children') as $key) {
     $new_item[$key] = $item[$key];
   }
diff --git a/modules/node/content_types.inc b/modules/node/content_types.inc
index ac37169..ac895cf 100644
--- a/modules/node/content_types.inc
+++ b/modules/node/content_types.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: content_types.inc,v 1.50.2.1 2008/02/13 11:23:28 goba Exp $
+// $Id: content_types.inc,v 1.50.2.2 2009/04/27 14:28:57 goba Exp $
 
 /**
  * @file
@@ -135,7 +135,7 @@ function node_type_form(&$form_state, $type = NULL) {
     '#type' => 'select',
     '#title' => t('Minimum number of words'),
     '#default_value' => $type->min_word_count,
-    '#options' => drupal_map_assoc(array(0, 10, 25, 50, 75, 100, 125, 150, 175, 200)),
+    '#options' => drupal_map_assoc(array(0, 1, 10, 25, 50, 75, 100, 125, 150, 175, 200)),
     '#description' => t('The minimum number of words for the body field to be considered valid for this content type. This can be useful to rule out submissions that do not meet the site\'s standards, such as short test posts.')
   );
   $form['submission']['help']  = array(
diff --git a/modules/node/node.info b/modules/node/node.info
index dd63c9f..5b0e067 100644
--- a/modules/node/node.info
+++ b/modules/node/node.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/node/node.pages.inc b/modules/node/node.pages.inc
index 430a5ec..72decf3 100644
--- a/modules/node/node.pages.inc
+++ b/modules/node/node.pages.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: node.pages.inc,v 1.28.2.3 2008/11/10 10:18:54 goba Exp $
+// $Id: node.pages.inc,v 1.28.2.4 2009/04/27 11:35:01 goba Exp $
 
 /**
  * @file
@@ -79,6 +79,8 @@ function node_object_prepare(&$node) {
   }
   else {
     $node->date = format_date($node->created, 'custom', 'Y-m-d H:i:s O');
+    // Remove the log message from the original node object.
+    $node->log = NULL;
   }
   // Always use the default revision setting.
   $node->revision = in_array('revision', $node_options);
@@ -159,6 +161,7 @@ function node_form(&$form_state, $node) {
     $form['revision_information']['log'] = array(
       '#type' => 'textarea',
       '#title' => t('Log message'),
+      '#default_value' => (isset($node->log) ? $node->log : ''),
       '#rows' => 2,
       '#description' => t('An explanation of the additions or updates being made to help other authors understand your motivations.'),
     );
diff --git a/modules/openid/openid.css b/modules/openid/openid.css
index 782f503..8c2dfc2 100644
--- a/modules/openid/openid.css
+++ b/modules/openid/openid.css
@@ -1,4 +1,4 @@
-/* $Id: openid.css,v 1.5 2008/01/30 22:11:22 goba Exp $ */
+/* $Id: openid.css,v 1.5.2.1 2009/03/31 10:48:22 dries Exp $ */
 
 #edit-openid-identifier {
   background-image: url("login-bg.png");
@@ -6,38 +6,35 @@
   background-repeat: no-repeat;
   padding-left: 20px;
 }
-
 div#edit-openid-identifier-wrapper {
   display: block;
 }
-
 html.js #user-login-form div#edit-openid-identifier-wrapper,
 html.js #user-login div#edit-openid-identifier-wrapper {
   display: none;
 }
-
 html.js #user-login-form li.openid-link,
 html.js #user-login li.openid-link {
   display : block;
+  list-style: none;
 }
-
 #user-login-form ul {
   margin-top: 0;
 }
-
+#user-login ul {
+  margin: 0 0 5px;
+}
+#user-login ul li {
+  margin: 0;
+}
 #user-login-form li.openid-link,
 #user-login-form li.user-link,
 #user-login li.openid-link,
 #user-login li.user-link {
   display: none;
 }
-
-#user-login-form li.openid-link,
-#user-login-form li.user-link {
-  text-align : left;
-}
-
-#user-login-form li.openid-link,
-#user-login li.openid-link {
-  background: transparent url(login-bg.png) no-repeat scroll 1px 0.35em;
-}
+#user-login-form li.openid-link a, 
+#user-login li.openid-link a {
+  background: transparent url("login-bg.png") no-repeat 0 2px;
+  padding: 0 20px;
+}
\ No newline at end of file
diff --git a/modules/openid/openid.info b/modules/openid/openid.info
index 96736f0..af8f0b1 100644
--- a/modules/openid/openid.info
+++ b/modules/openid/openid.info
@@ -5,8 +5,8 @@ version = VERSION
 package = Core - optional
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/openid/openid.module b/modules/openid/openid.module
index 8a21c52..4d84c55 100644
--- a/modules/openid/openid.module
+++ b/modules/openid/openid.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: openid.module,v 1.19.2.4 2009/01/14 21:36:16 goba Exp $
+// $Id: openid.module,v 1.19.2.5 2009/03/30 11:36:47 goba Exp $
 
 /**
  * @file
@@ -43,8 +43,7 @@ function openid_menu() {
  */
 function openid_help($path, $arg) {
   switch ($path) {
-
-  case 'user/%/openid':
+    case 'user/%/openid':
       $output = '<p>'. t('This site supports <a href="@openid-net">OpenID</a>, a secure way to log into many websites using a single username and password. OpenID can reduce the necessity of managing many usernames and passwords for many websites.', array('@openid-net' => url('http://openid.net'))) .'</p>';
       $output .= '<p>'. t('To use OpenID you must first establish an identity on a public or private OpenID server. If you do not have an OpenID and would like one, look into one of the <a href="@openid-providers">free public providers</a>. You can find out more about OpenID at <a href="@openid-net">this website</a>.', array('@openid-providers' => url('http://openid.net/wiki/index.php/OpenIDServers'), '@openid-net' => url('http://openid.net'))) .'</p>';
       $output .= '<p>'. t('If you already have an OpenID, enter the URL to your OpenID server below (e.g. myusername.openidprovider.com). Next time you login, you will be able to use this URL instead of a regular username and password. You can have multiple OpenID servers if you like; just keep adding them here.') .'</p>';
diff --git a/modules/path/path.info b/modules/path/path.info
index e05299f..15c0381 100644
--- a/modules/path/path.info
+++ b/modules/path/path.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/php/php.info b/modules/php/php.info
index 24873cf..9e11028 100644
--- a/modules/php/php.info
+++ b/modules/php/php.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/ping/ping.info b/modules/ping/ping.info
index ba431eb..4eca538 100644
--- a/modules/ping/ping.info
+++ b/modules/ping/ping.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/poll/poll.info b/modules/poll/poll.info
index e757320..93eea87 100644
--- a/modules/poll/poll.info
+++ b/modules/poll/poll.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/profile/profile.info b/modules/profile/profile.info
index d0a320b..ed17b35 100644
--- a/modules/profile/profile.info
+++ b/modules/profile/profile.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/search/search.info b/modules/search/search.info
index cf170e4..58cb8e4 100644
--- a/modules/search/search.info
+++ b/modules/search/search.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/search/search.module b/modules/search/search.module
index 358899f..ea9d8b5 100644
--- a/modules/search/search.module
+++ b/modules/search/search.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: search.module,v 1.250.2.5 2009/02/25 16:17:40 goba Exp $
+// $Id: search.module,v 1.250.2.6 2009/03/10 17:20:01 goba Exp $
 
 /**
  * @file
@@ -640,6 +640,7 @@ function search_nodeapi(&$node, $op, $teaser = NULL, $page = NULL) {
       if (count($output)) {
         return '<a>('. implode(', ', $output) .')</a>';
       }
+      break;
     // Reindex the node when it is updated.  The node is automatically indexed
     // when it is added, simply by being added to the node table.
     case 'update':
diff --git a/modules/statistics/statistics.info b/modules/statistics/statistics.info
index d654b6c..d07fab7 100644
--- a/modules/statistics/statistics.info
+++ b/modules/statistics/statistics.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/statistics/statistics.module b/modules/statistics/statistics.module
index b97662c..fa5fce1 100644
--- a/modules/statistics/statistics.module
+++ b/modules/statistics/statistics.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: statistics.module,v 1.272 2008/01/04 09:31:48 goba Exp $
+// $Id: statistics.module,v 1.272.2.1 2009/04/27 12:25:24 goba Exp $
 
 /**
  * @file
@@ -79,7 +79,7 @@ function statistics_link($type, $node = NULL, $teaser = FALSE) {
   global $id;
   $links = array();
 
-  if ($type != 'comment' && user_access('view post access counter')) {
+  if ($type == 'node' && user_access('view post access counter')) {
     $statistics = statistics_get($node->nid);
     if ($statistics) {
       $links['statistics_counter']['title'] = format_plural($statistics['totalcount'], '1 read', '@count reads');
diff --git a/modules/statistics/statistics.pages.inc b/modules/statistics/statistics.pages.inc
index 24f168f..9b0633e 100644
--- a/modules/statistics/statistics.pages.inc
+++ b/modules/statistics/statistics.pages.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: statistics.pages.inc,v 1.2 2007/10/20 21:57:50 goba Exp $
+// $Id: statistics.pages.inc,v 1.2.2.1 2009/04/27 12:08:29 goba Exp $
 
 /**
  * @file
@@ -15,7 +15,7 @@ function statistics_node_tracker() {
         array('data' => t('User'), 'field' => 'u.name'),
         array('data' => t('Operations')));
 
-    $result = pager_query('SELECT a.aid, a.timestamp, a.url, a.uid, u.name FROM {accesslog} a LEFT JOIN {users} u ON a.uid = u.uid WHERE a.path LIKE \'node/%d%%\''. tablesort_sql($header), 30, 0, NULL, $node->nid);
+    $result = pager_query("SELECT a.aid, a.timestamp, a.url, a.uid, u.name FROM {accesslog} a LEFT JOIN {users} u ON a.uid = u.uid WHERE a.path = 'node/%d' OR a.path LIKE 'node/%d/%%'". tablesort_sql($header), 30, 0, NULL, $node->nid, $node->nid);
     $rows = array();
     while ($log = db_fetch_object($result)) {
       $rows[] = array(
diff --git a/modules/syslog/syslog.info b/modules/syslog/syslog.info
index 0799538..2a65eba 100644
--- a/modules/syslog/syslog.info
+++ b/modules/syslog/syslog.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/system/maintenance-page.tpl.php b/modules/system/maintenance-page.tpl.php
index 5913aa2..5360f62 100644
--- a/modules/system/maintenance-page.tpl.php
+++ b/modules/system/maintenance-page.tpl.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: maintenance-page.tpl.php,v 1.2 2008/01/24 09:42:51 goba Exp $
+// $Id: maintenance-page.tpl.php,v 1.2.2.1 2009/04/30 00:13:31 goba Exp $
 
 /**
  * @file maintenance-page.tpl.php
@@ -19,8 +19,8 @@
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="<?php print $language->language ?>" lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
 
 <head>
-  <title><?php print $head_title; ?></title>
   <?php print $head; ?>
+  <title><?php print $head_title; ?></title>
   <?php print $styles; ?>
   <?php print $scripts; ?>
   <script type="text/javascript"><?php /* Needed to avoid Flash of Unstyled Content in IE */ ?> </script>
diff --git a/modules/system/page.tpl.php b/modules/system/page.tpl.php
index 817ec52..cb0a937 100644
--- a/modules/system/page.tpl.php
+++ b/modules/system/page.tpl.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: page.tpl.php,v 1.11 2008/01/24 09:42:51 goba Exp $
+// $Id: page.tpl.php,v 1.11.2.1 2009/04/30 00:13:31 goba Exp $
 
 /**
  * @file page.tpl.php
@@ -80,8 +80,8 @@
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="<?php print $language->language ?>" lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
 
 <head>
-  <title><?php print $head_title; ?></title>
   <?php print $head; ?>
+  <title><?php print $head_title; ?></title>
   <?php print $styles; ?>
   <?php print $scripts; ?>
   <script type="text/javascript"><?php /* Needed to avoid Flash of Unstyled Content in IE */ ?> </script>
diff --git a/modules/system/system.info b/modules/system/system.info
index 72b040f..ae93b73 100644
--- a/modules/system/system.info
+++ b/modules/system/system.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/system/system.install b/modules/system/system.install
index 8f689c9..91dbc1a 100644
--- a/modules/system/system.install
+++ b/modules/system/system.install
@@ -1,5 +1,5 @@
 <?php
-// $Id: system.install,v 1.238.2.12 2009/02/25 14:02:46 goba Exp $
+// $Id: system.install,v 1.238.2.14 2009/04/27 12:50:13 goba Exp $
 
 /**
  * Test and report Drupal installation requirements.
@@ -2526,6 +2526,15 @@ function system_update_6047() {
 }
 
 /**
+ * @} End of "defgroup updates-5.x-to-6.x"
+ */
+
+/**
+ * @defgroup updates-6.x-extra Extra system updates for 6.x
+ * @{
+ */
+
+/**
 * Increase the size of the 'load_functions' and 'to_arg_functions' fields in table 'menu_router'.
 */
 function system_update_6048() {
@@ -2536,7 +2545,6 @@ function system_update_6048() {
   return $ret;
 }
 
-
 /**
  * Replace src index on the {url_alias} table with src, language.
  */
@@ -2548,6 +2556,16 @@ function system_update_6049() {
 }
 
 /**
- * @} End of "defgroup updates-5.x-to-6.x"
+ * Clear any menu router blobs stored in the cache table.
+ */
+function system_update_6050() {
+  $ret = array();
+  cache_clear_all('router:', 'cache_menu', TRUE);
+  return $ret;
+}
+
+/**
+ * @} End of "defgroup updates-6.x-extra"
  * The next series of updates should start at 7000.
  */
+
diff --git a/modules/system/system.module b/modules/system/system.module
index f58fdb2..ad3e517 100644
--- a/modules/system/system.module
+++ b/modules/system/system.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: system.module,v 1.585.2.32 2009/02/25 21:02:37 goba Exp $
+// $Id: system.module,v 1.585.2.34 2009/04/30 00:13:31 goba Exp $
 
 /**
  * @file
@@ -9,7 +9,7 @@
 /**
  * The current system version.
  */
-define('VERSION', '6.10');
+define('VERSION', '6.11');
 
 /**
  * Core API compatibility.
diff --git a/modules/taxonomy/taxonomy.info b/modules/taxonomy/taxonomy.info
index 8d01b12..31f6eab 100644
--- a/modules/taxonomy/taxonomy.info
+++ b/modules/taxonomy/taxonomy.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/taxonomy/taxonomy.module b/modules/taxonomy/taxonomy.module
index 09d3a20..4e51f6b 100644
--- a/modules/taxonomy/taxonomy.module
+++ b/modules/taxonomy/taxonomy.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: taxonomy.module,v 1.414.2.6 2009/02/16 10:44:09 goba Exp $
+// $Id: taxonomy.module,v 1.414.2.8 2009/04/27 11:49:05 goba Exp $
 
 /**
  * @file
@@ -96,7 +96,6 @@ function taxonomy_link($type, $node = NULL) {
  * @return
  *   An internal Drupal path.
  */
-
 function taxonomy_term_path($term) {
   $vocabulary = taxonomy_vocabulary_load($term->vid);
   if ($vocabulary->module != 'taxonomy' && $path = module_invoke($vocabulary->module, 'term_path', $term)) {
@@ -1226,7 +1225,7 @@ function taxonomy_rss_item($node) {
   $output = array();
   foreach ($node->taxonomy as $term) {
     $output[] = array('key'   => 'category',
-                      'value' => check_plain($term->name),
+                      'value' => $term->name,
                       'attributes' => array('domain' => url('taxonomy/term/'. $term->tid, array('absolute' => TRUE))));
   }
   return $output;
diff --git a/modules/taxonomy/taxonomy.pages.inc b/modules/taxonomy/taxonomy.pages.inc
index 88fa185..e91459d 100644
--- a/modules/taxonomy/taxonomy.pages.inc
+++ b/modules/taxonomy/taxonomy.pages.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: taxonomy.pages.inc,v 1.9 2008/01/18 16:23:57 goba Exp $
+// $Id: taxonomy.pages.inc,v 1.9.2.1 2009/04/27 11:49:05 goba Exp $
 
 /**
  * @file
@@ -25,8 +25,8 @@ function taxonomy_term_page($str_tids = '', $depth = 0, $op = 'page') {
     }
 
     if ($names) {
-      $title = check_plain(implode(', ', $names));
-      drupal_set_title($title);
+      $title = implode(', ', $names);
+      drupal_set_title(check_plain($title));
 
       switch ($op) {
         case 'page':
diff --git a/modules/throttle/throttle.info b/modules/throttle/throttle.info
index 5879300..6b0afe9 100644
--- a/modules/throttle/throttle.info
+++ b/modules/throttle/throttle.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/tracker/tracker.info b/modules/tracker/tracker.info
index 49db2f2..2e6187d 100644
--- a/modules/tracker/tracker.info
+++ b/modules/tracker/tracker.info
@@ -6,8 +6,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/translation/translation.info b/modules/translation/translation.info
index 86b9e9f..1db3897 100644
--- a/modules/translation/translation.info
+++ b/modules/translation/translation.info
@@ -6,8 +6,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/trigger/trigger.info b/modules/trigger/trigger.info
index f9d172a..9474fbb 100644
--- a/modules/trigger/trigger.info
+++ b/modules/trigger/trigger.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/trigger/trigger.module b/modules/trigger/trigger.module
index 0358f9c..718a991 100644
--- a/modules/trigger/trigger.module
+++ b/modules/trigger/trigger.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: trigger.module,v 1.13.2.1 2008/04/09 21:11:51 goba Exp $
+// $Id: trigger.module,v 1.13.2.2 2009/03/30 11:53:09 goba Exp $
 
 /**
  * @file
@@ -240,7 +240,7 @@ function trigger_nodeapi(&$node, $op, $a3, $a4) {
       }
       // Since we know about the node, we pass that info along to the action.
       $context['node'] = $node;
-      $result = actions_do($aid, $objects[$action_info['type']], $context, $a4, $a4);
+      $result = actions_do($aid, $objects[$action_info['type']], $context, $a3, $a4);
     }
     else {
       actions_do($aid, $node, $context, $a3, $a4);
diff --git a/modules/update/update.compare.inc b/modules/update/update.compare.inc
index e2fd777..ce7aa2f 100644
--- a/modules/update/update.compare.inc
+++ b/modules/update/update.compare.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.compare.inc,v 1.8.2.2 2008/08/28 08:14:56 dries Exp $
+// $Id: update.compare.inc,v 1.8.2.3 2009/04/29 18:43:11 goba Exp $
 
 /**
  * @file
@@ -16,9 +16,18 @@
  * that logic is only required when preparing the status report, not for
  * fetching the available release data.
  *
+ * This array is fairly expensive to construct, since it involves a lot of
+ * disk I/O, so we cache the results into the {cache_update} table using the
+ * 'update_project_projects' cache ID. However, since this is not the data
+ * about available updates fetched from the network, it is ok to invalidate it
+ * somewhat quickly. If we keep this data for very long, site administrators
+ * are more likely to see incorrect results if they upgrade to a newer version
+ * of a module or theme but do not visit certain pages that automatically
+ * clear this cache.
+ *
  * @see update_process_project_info()
  * @see update_calculate_project_data()
- *
+ * @see update_project_cache()
  */
 function update_get_projects() {
   static $projects = array();
@@ -29,8 +38,8 @@ function update_get_projects() {
       // Still empty, so we have to rebuild the cache.
       _update_process_info_list($projects, module_rebuild_cache(), 'module');
       _update_process_info_list($projects, system_theme_data(), 'theme');
-      // Set the projects array into the cache table.
-      cache_set('update_project_projects', $projects, 'cache_update', time() + 3600);
+      // Cache the site's project data for at most 1 hour.
+      _update_cache_set('update_project_projects', $projects, time() + 3600);
     }
   }
   return $projects;
@@ -223,12 +232,23 @@ function update_process_project_info(&$projects) {
  * version (e.g. 5.x-1.5-beta1, 5.x-1.5-beta2, and 5.x-1.5). Development
  * snapshots for a given major version are always listed last.
  *
+ * The results of this function are expensive to compute, especially on sites
+ * with lots of modules or themes, since it involves a lot of comparisons and
+ * other operations. Therefore, we cache the results into the {cache_update}
+ * table using the 'update_project_data' cache ID. However, since this is not
+ * the data about available updates fetched from the network, it is ok to
+ * invalidate it somewhat quickly. If we keep this data for very long, site
+ * administrators are more likely to see incorrect results if they upgrade to
+ * a newer version of a module or theme but do not visit certain pages that
+ * automatically clear this cache.
+ *
  * @param $available
  *  Array of data about available project releases.
  *
  * @see update_get_available()
  * @see update_get_projects()
  * @see update_process_project_info()
+ * @see update_project_cache()
  */
 function update_calculate_project_data($available) {
   // Retrieve the projects from cache, if present.
@@ -444,7 +464,7 @@ function update_calculate_project_data($available) {
 
         // If we're running a dev snapshot and have a timestamp, stop
         // searching for security updates once we hit an official release
-        // older than what we've got.  Allow 100 seconds of leeway to handle
+        // older than what we've got. Allow 100 seconds of leeway to handle
         // differences between the datestamp in the .info file and the
         // timestamp of the tarball itself (which are usually off by 1 or 2
         // seconds) so that we don't flag that as a new release.
@@ -550,8 +570,8 @@ function update_calculate_project_data($available) {
   // projects or releases).
   drupal_alter('update_status', $projects);
 
-  // Set the projects array into the cache table.
-  cache_set('update_project_data', $projects, 'cache_update', time() + 3600);
+  // Cache the site's update status for at most 1 hour.
+  _update_cache_set('update_project_data', $projects, time() + 3600);
   return $projects;
 }
 
@@ -567,6 +587,13 @@ function update_calculate_project_data($available) {
  * administration pages, since we should always recompute the most current
  * values on any of those pages.
  *
+ * Note: while both of these arrays are expensive to compute (in terms of disk
+ * I/O and some fairly heavy CPU processing), neither of these is the actual
+ * data about available updates that we have to fetch over the network from
+ * updates.drupal.org. That information is stored with the
+ * 'update_available_releases' cache ID -- it needs to persist longer than 1
+ * hour and never get invalidated just by visiting a page on the site.
+ *
  * @param $cid
  *   The cache id of data to return from the cache. Valid options are
  *   'update_project_data' and 'update_project_projects'.
@@ -579,15 +606,15 @@ function update_calculate_project_data($available) {
 function update_project_cache($cid) {
   $projects = array();
 
-  // In some cases, we must clear the cache.  Rather than do so on a time
-  // basis, we check for specific paths.
+  // On certain paths, we should clear the cache and recompute the projects or
+  // update status of the site to avoid presenting stale information.
   $q = $_GET['q'];
   $paths = array('admin/build/modules', 'admin/build/themes', 'admin/reports', 'admin/reports/updates', 'admin/reports/status', 'admin/reports/updates/check');
   if (in_array($q, $paths)) {
-    cache_clear_all($cid, 'cache_update');
+    _update_cache_clear($cid);
   }
   else {
-    $cache = cache_get($cid, 'cache_update');
+    $cache = _update_cache_get($cid);
     if (!empty($cache->data) && $cache->expire > time()) {
       $projects = $cache->data;
     }
diff --git a/modules/update/update.css b/modules/update/update.css
index d797f0e..14595af 100644
--- a/modules/update/update.css
+++ b/modules/update/update.css
@@ -1,4 +1,4 @@
-/* $Id: update.css,v 1.3.2.1 2008/02/05 09:59:21 goba Exp $ */
+/* $Id: update.css,v 1.3.2.2 2009/04/29 17:17:20 goba Exp $ */
 
 .update .project {
   font-weight: bold;
@@ -56,6 +56,10 @@
   direction: ltr; /* Note: version numbers should always be LTR. */
 }
 
+.update tr.unknown {
+  background: #ddd;
+}
+
 table.update,
 .update table.version {
   width: 100%;
diff --git a/modules/update/update.fetch.inc b/modules/update/update.fetch.inc
index 1bb6e04..22db6e8 100644
--- a/modules/update/update.fetch.inc
+++ b/modules/update/update.fetch.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.fetch.inc,v 1.7.2.2 2009/01/14 21:36:16 goba Exp $
+// $Id: update.fetch.inc,v 1.7.2.4 2009/04/29 18:43:11 goba Exp $
 
 /**
  * @file
@@ -27,17 +27,24 @@ function _update_refresh() {
   module_load_include('inc', 'update', 'update.compare');
 
   // Since we're fetching new available update data, we want to clear
-  // everything in our cache, to ensure we recompute the status. Note that
-  // this does not cause update_get_projects() to be recomputed twice in the
-  // same page load (e.g. when manually checking) since that function stashes
-  // its answer in a static array.
-  update_invalidate_cache();
+  // our cache of both the projects we care about, and the current update
+  // status of the site. We do *not* want to clear the cache of available
+  // releases just yet, since that data (even if it's stale) can be useful
+  // during update_get_projects(); for example, to modules that implement
+  // hook_system_info_alter() such as cvs_deploy.
+  _update_cache_clear('update_project_projects');
+  _update_cache_clear('update_project_data');
 
   $available = array();
   $data = array();
   $site_key = md5($base_url . drupal_get_private_key());
   $projects = update_get_projects();
 
+  // Now that we have the list of projects, we should also clear our cache of
+  // available release data, since even if we fail to fetch new data, we need
+  // to clear out the stale data at this point.
+  _update_cache_clear('update_available_releases');
+  
   foreach ($projects as $key => $project) {
     $url = _update_build_fetch_url($project, $site_key);
     $xml = drupal_http_request($url);
@@ -52,7 +59,7 @@ function _update_refresh() {
   }
   if (!empty($available) && is_array($available)) {
     $frequency = variable_get('update_check_frequency', 1);
-    cache_set('update_info', $available, 'cache_update', time() + (60 * 60 * 24 * $frequency));
+    _update_cache_set('update_available_releases', $available, time() + (60 * 60 * 24 * $frequency));
     variable_set('update_last_check', time());
     watchdog('update', 'Fetched information about all available new releases and updates.', array(), WATCHDOG_NOTICE, l(t('view'), 'admin/reports/updates'));
   }
@@ -101,7 +108,7 @@ function _update_build_fetch_url($project, $site_key = '') {
  * Perform any notifications that should be done once cron fetches new data.
  *
  * This method checks the status of the site using the new data and depending
- * on the configuration of the site, notifys administrators via email if there
+ * on the configuration of the site, notifies administrators via email if there
  * are new releases or missing security updates.
  *
  * @see update_requirements()
@@ -110,10 +117,11 @@ function _update_cron_notify() {
   include_once './includes/install.inc';
   $status = update_requirements('runtime');
   $params = array();
+  $notify_all = (variable_get('update_notification_threshold', 'all') == 'all');
   foreach (array('core', 'contrib') as $report_type) {
     $type = 'update_'. $report_type;
     if (isset($status[$type]['severity'])
-        && $status[$type]['severity'] == REQUIREMENT_ERROR) {
+        && ($status[$type]['severity'] == REQUIREMENT_ERROR || ($notify_all && $status[$type]['reason'] == UPDATE_NOT_CURRENT))) {
       $params[$report_type] = $status[$type]['reason'];
     }
   }
diff --git a/modules/update/update.info b/modules/update/update.info
index b73bee8..f0db1b6 100644
--- a/modules/update/update.info
+++ b/modules/update/update.info
@@ -5,8 +5,8 @@ version = VERSION
 package = Core - optional
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/update/update.module b/modules/update/update.module
index c6f2653..a1a23fb 100644
--- a/modules/update/update.module
+++ b/modules/update/update.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.module,v 1.17.2.1 2008/09/23 10:19:02 goba Exp $
+// $Id: update.module,v 1.17.2.3 2009/04/29 18:43:11 goba Exp $
 
 /**
  * @file
@@ -168,7 +168,7 @@ function update_theme() {
 }
 
 /**
- * Implementation of hook_requirements.
+ * Implementation of hook_requirements().
  *
  * @return
  *   An array describing the status of the site regarding available updates.
@@ -263,7 +263,7 @@ function _update_requirement_check($project, $type) {
       break;
     case UPDATE_NOT_CURRENT:
       $requirement_label = t('Out of date');
-      $requirement['severity'] = variable_get('update_notification_threshold', 'all') == 'all' ? REQUIREMENT_ERROR : REQUIREMENT_WARNING;
+      $requirement['severity'] = REQUIREMENT_WARNING;
       break;
     case UPDATE_UNKNOWN:
     case UPDATE_NOT_CHECKED:
@@ -286,9 +286,9 @@ function _update_requirement_check($project, $type) {
 function update_cron() {
   $frequency = variable_get('update_check_frequency', 1);
   $interval = 60 * 60 * 24 * $frequency;
-  // Cron should check for updates if there is no update data cached or if the configured
-  // update interval has elapsed.
-  if (!cache_get('update_info', 'cache_update') || ((time() - variable_get('update_last_check', 0)) > $interval)) {
+  // Cron should check for updates if there is no update data cached or if the
+  // configured update interval has elapsed.
+  if (!_update_cache_get('update_available_releases') || ((time() - variable_get('update_last_check', 0)) > $interval)) {
     update_refresh();
     _update_cron_notify();
   }
@@ -353,8 +353,7 @@ function update_get_available($refresh = FALSE) {
       break;
     }
   }
-  if (!$needs_refresh && ($cache = cache_get('update_info', 'cache_update'))
-       && $cache->expire > time()) {
+  if (!$needs_refresh && ($cache = _update_cache_get('update_available_releases')) && $cache->expire > time()) {
     $available = $cache->data;
   }
   elseif ($needs_refresh || $refresh) {
@@ -367,24 +366,6 @@ function update_get_available($refresh = FALSE) {
 }
 
 /**
- * Implementation of hook_flush_caches().
- *
- * The function update.php (among others) calls this hook to flush the caches.
- * Since we're running update.php, we are likely to install a new version of
- * something, in which case, we want to check for available update data again.
- */
-function update_flush_caches() {
-  return array('cache_update');
-}
-
-/**
- * Invalidates any cached data relating to update status.
- */
-function update_invalidate_cache() {
-  cache_clear_all('*', 'cache_update', TRUE);
-}
-
-/**
  * Wrapper to load the include file and then refresh the release data.
  */
 function update_refresh() {
@@ -514,3 +495,122 @@ function _update_project_status_sort($a, $b) {
   $b_status = $b['status'] > 0 ? $b['status'] : (-10 * $b['status']);
   return $a_status - $b_status;
 }
+
+/**
+ * @defgroup update_status_cache Private update status cache system
+ * @{
+ *
+ * We specifically do NOT use the core cache API for saving the fetched data
+ * about available updates. It is vitally important that this cache is only
+ * cleared when we're populating it after successfully fetching new available
+ * update data. Usage of the core cache API results in all sorts of potential
+ * problems that would result in attempting to fetch available update data all
+ * the time, including if a site has a "minimum cache lifetime" (which is both
+ * a minimum and a maximum) defined, or if a site uses memcache or another
+ * plug-able cache system that assumes volatile caches.
+ *
+ * Update module still uses the {cache_update} table, but instead of using
+ * cache_set(), cache_get(), and cache_clear_all(), there are private helper
+ * functions that implement these same basic tasks but ensure that the cache
+ * is not prematurely cleared, and that the data is always stored in the
+ * database, even if memcache or another cache backend is in use.
+ */
+
+/**
+ * Store data in the private update status cache table.
+ *
+ * Note: this function completely ignores the {cache_update}.headers field
+ * since that is meaningless for the kinds of data we're caching.
+ *
+ * @param $cid
+ *   The cache ID to save the data with.
+ * @param $data
+ *   The data to store.
+ * @param $expire
+ *   One of the following values:
+ *   - CACHE_PERMANENT: Indicates that the item should never be removed except
+ *     by explicitly using _update_cache_clear() or update_invalidate_cache().
+ *   - A Unix timestamp: Indicates that the item should be kept at least until
+ *     the given time, after which it will be invalidated.
+ */
+function _update_cache_set($cid, $data, $expire) {
+  $serialized = 0;
+  if (is_object($data) || is_array($data)) {
+    $data = serialize($data);
+    $serialized = 1;
+  }
+  $created = time();
+  db_query("UPDATE {cache_update} SET data = %b, created = %d, expire = %d, serialized = %d WHERE cid = '%s'", $data, $created, $expire, $serialized, $cid);
+  if (!db_affected_rows()) {
+    @db_query("INSERT INTO {cache_update} (cid, data, created, expire, serialized) VALUES ('%s', %b, %d, %d, %d)", $cid, $data, $created, $expire, $serialized);
+  }
+}
+
+/** 
+ * Retrieve data from the private update status cache table.
+ *
+ * @param $cid
+ *   The cache ID to retrieve.
+ * @return
+ *   The data for the given cache ID, or NULL if the ID was not found.
+ */
+function _update_cache_get($cid) {
+  $cache = db_fetch_object(db_query("SELECT data, created, expire, serialized FROM {cache_update} WHERE cid = '%s'", $cid));
+  if (isset($cache->data)) {
+    $cache->data = db_decode_blob($cache->data);
+    if ($cache->serialized) {
+      $cache->data = unserialize($cache->data);
+    }
+  }
+  return $cache;
+}
+
+/**
+ * Invalidates specific cached data relating to update status.
+ *
+ * @param $cid
+ *   Optional cache ID of the record to clear from the private update module
+ *   cache. If empty, all records will be cleared from the table.
+ */
+function _update_cache_clear($cid = NULL) {
+  if (empty($cid)) {
+    db_query("DELETE FROM {cache_update}");
+  }
+  else {
+    db_query("DELETE FROM {cache_update} WHERE cid = '%s'", $cid);
+  }
+}
+
+/**
+ * Implementation of hook_flush_caches().
+ *
+ * Called from update.php (among others) to flush the caches.
+ * Since we're running update.php, we are likely to install a new version of
+ * something, in which case, we want to check for available update data again.
+ * However, because we have our own caching system, we need to directly clear
+ * the database table ourselves at this point and return nothing, for example,
+ * on sites that use memcache where cache_clear_all() won't know how to purge
+ * this data.
+ *
+ * However, we only want to do this from update.php, since otherwise, we'd
+ * lose all the available update data on every cron run. So, we specifically
+ * check if the site is in MAINTENANCE_MODE == 'update' (which indicates
+ * update.php is running, not update module... alas for overloaded names).
+ */
+function update_flush_caches() {
+  if (defined('MAINTENANCE_MODE') && MAINTENANCE_MODE == 'update') {
+    _update_cache_clear();
+  }
+  return array();
+}
+
+/**
+ * Invalidates all cached data relating to update status.
+ */
+function update_invalidate_cache() {
+  _update_cache_clear();
+}
+
+/**
+ * @} End of "defgroup update_status_cache".
+ */
diff --git a/modules/update/update.report.inc b/modules/update/update.report.inc
index ec0c43f..55f823a 100644
--- a/modules/update/update.report.inc
+++ b/modules/update/update.report.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.report.inc,v 1.10.2.2 2008/08/28 08:14:56 dries Exp $
+// $Id: update.report.inc,v 1.10.2.3 2009/04/29 17:17:21 goba Exp $
 
 /**
  * @file
@@ -47,17 +47,18 @@ function theme_update_report($data) {
         $class = 'ok';
         $icon = theme('image', 'misc/watchdog-ok.png', t('ok'), t('ok'));
         break;
+      case UPDATE_UNKNOWN:
+        $class = 'unknown';
+        $icon = theme('image', 'misc/watchdog-warning.png', t('warning'), t('warning'));
+        break;
       case UPDATE_NOT_SECURE:
       case UPDATE_REVOKED:
       case UPDATE_NOT_SUPPORTED:
+        $class = 'error';
+        $icon = theme('image', 'misc/watchdog-error.png', t('error'), t('error'));
+        break;
+      case UPDATE_NOT_CHECKED:
       case UPDATE_NOT_CURRENT:
-        if ($notification_level == 'all'
-            || $project['status'] != UPDATE_NOT_CURRENT) {
-          $class = 'error';
-          $icon = theme('image', 'misc/watchdog-error.png', t('error'), t('error'));
-          break;
-        }
-        // Otherwise, deliberate no break and use the warning class/icon.
       default:
         $class = 'warning';
         $icon = theme('image', 'misc/watchdog-warning.png', t('warning'), t('warning'));
diff --git a/modules/update/update.settings.inc b/modules/update/update.settings.inc
index d6f4f04..5eab866 100644
--- a/modules/update/update.settings.inc
+++ b/modules/update/update.settings.inc
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.settings.inc,v 1.3 2007/10/20 21:57:50 goba Exp $
+// $Id: update.settings.inc,v 1.3.2.1 2009/04/29 17:17:21 goba Exp $
 
 /**
  * @file
@@ -34,13 +34,13 @@ function update_settings() {
 
   $form['update_notification_threshold'] = array(
     '#type' => 'radios',
-    '#title' => t('Notification threshold'),
+    '#title' => t('E-mail notification threshold'),
     '#default_value' => variable_get('update_notification_threshold', 'all'),
     '#options' => array(
       'all' => t('All newer versions'),
       'security' => t('Only security updates'),
     ),
-    '#description' => t('If there are updates available of Drupal core or any of your installed modules and themes, your site will print an error message on the <a href="@status_report">status report</a>, the <a href="@modules_page">modules page</a>, and the <a href="@themes_page">themes page</a>. You can choose to only see these error messages if a security update is available, or to be notified about any newer versions.', array('@status_report' => url('admin/reports/status'), '@modules_page' => url('admin/build/modules'), '@themes_page' => url('admin/build/themes')))
+    '#description' => t('You can choose to send e-mail only if a security update is available, or to be notified about all newer versions. If there are updates available of Drupal core or any of your installed modules and themes, your site will always print a message on the <a href="@status_report">status report</a> page, and will also display an error message on administration pages if there is a security update.', array('@status_report' => url('admin/reports/status')))
   );
 
   $form = system_settings_form($form);
diff --git a/modules/upload/upload.info b/modules/upload/upload.info
index 48fb119..abc85c8 100644
--- a/modules/upload/upload.info
+++ b/modules/upload/upload.info
@@ -5,8 +5,8 @@ package = Core - optional
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/user/user.info b/modules/user/user.info
index 4044d91..8bb3a8f 100644
--- a/modules/user/user.info
+++ b/modules/user/user.info
@@ -5,8 +5,8 @@ package = Core - required
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/modules/user/user.module b/modules/user/user.module
index 5ada13d..373f05a 100644
--- a/modules/user/user.module
+++ b/modules/user/user.module
@@ -1,5 +1,5 @@
 <?php
-// $Id: user.module,v 1.892.2.12 2009/02/25 13:57:04 goba Exp $
+// $Id: user.module,v 1.892.2.13 2009/04/27 12:02:27 goba Exp $
 
 /**
  * @file
@@ -1249,7 +1249,6 @@ function user_login(&$form_state) {
     '#size' => 60,
     '#maxlength' => USERNAME_MAX_LENGTH,
     '#required' => TRUE,
-    '#attributes' => array('tabindex' => '1'),
   );
 
   $form['name']['#description'] = t('Enter your @s username.', array('@s' => variable_get('site_name', 'Drupal')));
@@ -1257,10 +1256,9 @@ function user_login(&$form_state) {
     '#title' => t('Password'),
     '#description' => t('Enter the password that accompanies your username.'),
     '#required' => TRUE,
-    '#attributes' => array('tabindex' => '2'),
   );
   $form['#validate'] = user_login_default_validators();
-  $form['submit'] = array('#type' => 'submit', '#value' => t('Log in'), '#weight' => 2, '#attributes' => array('tabindex' => '3'));
+  $form['submit'] = array('#type' => 'submit', '#value' => t('Log in'), '#weight' => 2);
 
   return $form;
 }
diff --git a/themes/bluemarine/bluemarine.info b/themes/bluemarine/bluemarine.info
index 8a392bb..98b7d47 100644
--- a/themes/bluemarine/bluemarine.info
+++ b/themes/bluemarine/bluemarine.info
@@ -5,8 +5,8 @@ version = VERSION
 core = 6.x
 engine = phptemplate
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/themes/bluemarine/page.tpl.php b/themes/bluemarine/page.tpl.php
index 368822b..258a229 100644
--- a/themes/bluemarine/page.tpl.php
+++ b/themes/bluemarine/page.tpl.php
@@ -1,11 +1,11 @@
 <?php
-// $Id: page.tpl.php,v 1.28 2008/01/24 09:42:52 goba Exp $
+// $Id: page.tpl.php,v 1.28.2.1 2009/04/30 00:13:31 goba Exp $
 ?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml" lang="<?php print $language->language ?>" xml:lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
 
 <head>
-  <title><?php print $head_title ?></title>
   <?php print $head ?>
+  <title><?php print $head_title ?></title>
   <?php print $styles ?>
   <?php print $scripts ?>
   <script type="text/javascript"><?php /* Needed to avoid Flash of Unstyle Content in IE */ ?> </script>
diff --git a/themes/chameleon/chameleon.info b/themes/chameleon/chameleon.info
index 121bac7..e2701fc 100644
--- a/themes/chameleon/chameleon.info
+++ b/themes/chameleon/chameleon.info
@@ -12,8 +12,8 @@ stylesheets[all][] = common.css
 version = VERSION
 core = 6.x
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/themes/chameleon/chameleon.theme b/themes/chameleon/chameleon.theme
index 7ce0d5e..b9fc26a 100644
--- a/themes/chameleon/chameleon.theme
+++ b/themes/chameleon/chameleon.theme
@@ -1,5 +1,5 @@
 <?php
-// $Id: chameleon.theme,v 1.76 2008/01/24 09:42:53 goba Exp $
+// $Id: chameleon.theme,v 1.76.2.1 2009/04/30 00:13:31 goba Exp $
 
 /**
  * @file
@@ -30,8 +30,8 @@ function chameleon_page($content, $show_blocks = TRUE, $show_messages = TRUE) {
   $output  = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n";
   $output .= "<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"$language\" xml:lang=\"$language\" dir=\"$direction\">\n";
   $output .= "<head>\n";
-  $output .= " <title>". ($title ? strip_tags($title) ." | ". variable_get("site_name", "Drupal") : variable_get("site_name", "Drupal") ." | ". variable_get("site_slogan", "")) ."</title>\n";
   $output .= drupal_get_html_head();
+  $output .= " <title>". ($title ? strip_tags($title) ." | ". variable_get("site_name", "Drupal") : variable_get("site_name", "Drupal") ." | ". variable_get("site_slogan", "")) ."</title>\n";
   $output .= drupal_get_css();
   $output .= drupal_get_js();
   $output .= "</head>";
diff --git a/themes/chameleon/marvin/marvin.info b/themes/chameleon/marvin/marvin.info
index 30c46ef..ab8a9c0 100644
--- a/themes/chameleon/marvin/marvin.info
+++ b/themes/chameleon/marvin/marvin.info
@@ -7,8 +7,8 @@ version = VERSION
 core = 6.x
 base theme = chameleon
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/themes/garland/garland.info b/themes/garland/garland.info
index d23b8e6..5275271 100644
--- a/themes/garland/garland.info
+++ b/themes/garland/garland.info
@@ -7,8 +7,8 @@ engine = phptemplate
 stylesheets[all][] = style.css
 stylesheets[print][] = print.css
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/themes/garland/maintenance-page.tpl.php b/themes/garland/maintenance-page.tpl.php
index de46abf..bda5bf4 100644
--- a/themes/garland/maintenance-page.tpl.php
+++ b/themes/garland/maintenance-page.tpl.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: maintenance-page.tpl.php,v 1.3 2008/01/24 09:42:53 goba Exp $
+// $Id: maintenance-page.tpl.php,v 1.3.2.1 2009/04/30 00:13:31 goba Exp $
 
 /**
  * @file maintenance-page.tpl.php
@@ -15,8 +15,8 @@
   "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="<?php print $language->language ?>" lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
   <head>
-    <title><?php print $head_title ?></title>
     <?php print $head ?>
+    <title><?php print $head_title ?></title>
     <?php print $styles ?>
     <?php print $scripts ?>
     <!--[if lt IE 7]>
diff --git a/themes/garland/minnelli/minnelli.info b/themes/garland/minnelli/minnelli.info
index 59d4afe..bf497ec 100644
--- a/themes/garland/minnelli/minnelli.info
+++ b/themes/garland/minnelli/minnelli.info
@@ -6,8 +6,8 @@ core = 6.x
 base theme = garland
 stylesheets[all][] = minnelli.css
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/themes/garland/page.tpl.php b/themes/garland/page.tpl.php
index 3e652ad..3f6cbe8 100644
--- a/themes/garland/page.tpl.php
+++ b/themes/garland/page.tpl.php
@@ -1,11 +1,11 @@
 <?php
-// $Id: page.tpl.php,v 1.18 2008/01/24 09:42:53 goba Exp $
+// $Id: page.tpl.php,v 1.18.2.1 2009/04/30 00:13:31 goba Exp $
 ?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
   "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="<?php print $language->language ?>" lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
   <head>
-    <title><?php print $head_title ?></title>
     <?php print $head ?>
+    <title><?php print $head_title ?></title>
     <?php print $styles ?>
     <?php print $scripts ?>
     <!--[if lt IE 7]>
diff --git a/themes/garland/style.css b/themes/garland/style.css
index f09cf5d..87de957 100644
--- a/themes/garland/style.css
+++ b/themes/garland/style.css
@@ -1,4 +1,4 @@
-/* $Id: style.css,v 1.38.2.1 2008/02/05 09:27:26 goba Exp $ */
+/* $Id: style.css,v 1.38.2.2 2009/04/27 14:20:53 goba Exp $ */
 
 /**
  * Garland, for Drupal 6.x
@@ -659,7 +659,7 @@ ul.secondary li.active a {
  */
 .node {
   border-bottom: 1px solid #e9eff3;
-  margin: -1.5em -26px 1.5em;
+  margin: 0 -26px 1.5em;
   padding: 1.5em 26px;
 }
 
diff --git a/themes/pushbutton/page.tpl.php b/themes/pushbutton/page.tpl.php
index d8b8a9e..a3867aa 100644
--- a/themes/pushbutton/page.tpl.php
+++ b/themes/pushbutton/page.tpl.php
@@ -1,11 +1,11 @@
 <?php
-// $Id: page.tpl.php,v 1.25.2.1 2008/09/29 13:32:15 goba Exp $
+// $Id: page.tpl.php,v 1.25.2.2 2009/04/30 00:13:31 goba Exp $
 ?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml" lang="<?php print $language->language ?>" xml:lang="<?php print $language->language ?>" dir="<?php print $language->dir ?>">
 <head>
-  <title><?php print $head_title ?></title>
   <meta http-equiv="Content-Style-Type" content="text/css" />
   <?php print $head ?>
+  <title><?php print $head_title ?></title>
   <?php print $styles ?>
   <?php print $scripts ?>
 </head>
diff --git a/themes/pushbutton/pushbutton.info b/themes/pushbutton/pushbutton.info
index c82ca7a..fb2ebb9 100644
--- a/themes/pushbutton/pushbutton.info
+++ b/themes/pushbutton/pushbutton.info
@@ -5,8 +5,8 @@ version = VERSION
 core = 6.x
 engine = phptemplate
 
-; Information added by drupal.org packaging script on 2009-02-25
-version = "6.10"
+; Information added by drupal.org packaging script on 2009-04-30
+version = "6.11"
 project = "drupal"
-datestamp = "1235596218"
+datestamp = "1241050838"
 
diff --git a/update.php b/update.php
index 6f9ba72..28b1b7a 100644
--- a/update.php
+++ b/update.php
@@ -1,5 +1,5 @@
 <?php
-// $Id: update.php,v 1.252.2.2 2008/12/10 22:30:13 goba Exp $
+// $Id: update.php,v 1.252.2.3 2009/03/30 11:15:11 goba Exp $
 
 /**
  * @file
@@ -247,7 +247,6 @@ function update_script_selection_form() {
   $form['has_js'] = array(
     '#type' => 'hidden',
     '#default_value' => FALSE,
-    '#attributes' => array('id' => 'edit-has_js'),
   );
   $form['submit'] = array(
     '#type' => 'submit',
@@ -378,7 +377,7 @@ function update_info_page() {
   $output .= "<li>Install your new files in the appropriate location, as described in the handbook.</li>\n";
   $output .= "</ol>\n";
   $output .= "<p>When you have performed the steps above, you may proceed.</p>\n";
-  $output .= '<form method="post" action="update.php?op=selection&token='. $token .'"><input type="submit" value="Continue" /></form>';
+  $output .= '<form method="post" action="update.php?op=selection&amp;token='. $token .'"><p><input type="submit" value="Continue" /></p></form>';
   $output .= "\n";
   return $output;
 }
