diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index 8f6f713..cc72f06 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,4 +1,8 @@
 
+Drupal 6.23, 2012-02-01
+----------------------
+- Fixed security issues (Cross site scripting), see SA-CORE-2012-001.
+
 Drupal 6.22, 2011-05-25
 ----------------------
 - Made Drupal 6 work better with IIS and Internet Explorer.
diff --git a/includes/database.inc b/includes/database.inc
index ad36f8e..64c67c6 100644
--- a/includes/database.inc
+++ b/includes/database.inc
@@ -679,3 +679,24 @@ function db_type_placeholder($type) {
 /**
  * @} End of "defgroup schemaapi".
  */
+
+/**
+ * Sets a session variable specifying the lag time for ignoring a slave server.
+ * Backport of D7 functionality.
+ */
+function db_ignore_slave() {
+  global $active_slave_db;
+
+  // Only set ignore_slave_server if there are slave servers being used, which
+  // is assumed db_url is an array.
+  if (isset($active_slave_db)) {
+    // Five minutes is long enough to allow the slave to break and resume
+    // interrupted replication without causing problems on the Drupal site from
+    // the old data.
+    $duration = variable_get('maximum_replication_lag', 300);
+    // Set session variable with amount of time to delay before using slave.
+    // This will stick around for 5 minutes by default. 
+    // There is another $_Session variable 'not_slavesafe' for single queries.
+    $_SESSION['ignore_slave_server'] = $_SERVER['REQUEST_TIME'] + $duration;
+  }
+}
diff --git a/includes/database.mysql.inc b/includes/database.mysql.inc
index ed4fb4f..acad975 100644
--- a/includes/database.mysql.inc
+++ b/includes/database.mysql.inc
@@ -98,6 +98,28 @@ function db_connect($url) {
 function _db_query($query, $debug = 0, $slave = FALSE) {
   global $active_db, $active_slave_db, $queries, $user;
 
+  // Ignore slave database servers for this request.
+  //
+  // In Drupal's distributed database structure, new data is written to the master
+  // and then propagated to the slave servers.  This means there is a lag
+  // between when data is written to the master and when it is available on the slave.
+  // At these times, we will want to avoid using a slave server temporarily.
+  // For example, if a user posts a new node then we want to disable the slave
+  // server for that user temporarily to allow the slave server to catch up.
+  // That way, that user will see their changes immediately while for other
+  // users we still get the benefits of having a slave server, just with slightly
+  // stale data.  Code that wants to disable the slave server should use the
+  // db_set_ignore_slave() function to set $_SESSION['ignore_slave_server'] to
+  // the timestamp after which the slave can be re-enabled.
+  if (isset($_SESSION['ignore_slave_server'])) {
+    if ($_SESSION['ignore_slave_server'] >= $_SERVER['REQUEST_TIME']) {
+      $slave = FALSE;
+    }
+    else {
+      unset($_SESSION['ignore_slave_server']);
+    }
+  }
+
   if (variable_get('dev_query', 0)) {
     list($usec, $sec) = explode(' ', microtime());
     $timer = (float)$usec + (float)$sec;
diff --git a/includes/database.mysqli.inc b/includes/database.mysqli.inc
index 61a1da6..5ac4ab9 100644
--- a/includes/database.mysqli.inc
+++ b/includes/database.mysqli.inc
@@ -97,6 +97,28 @@ function db_connect($url) {
 function _db_query($query, $debug = 0, $slave = FALSE) {
   global $active_db, $active_slave_db, $queries, $user;
 
+  // Ignore slave database servers for this request.
+  //
+  // In Drupal's distributed database structure, new data is written to the master
+  // and then propagated to the slave servers.  This means there is a lag
+  // between when data is written to the master and when it is available on the slave.
+  // At these times, we will want to avoid using a slave server temporarily.
+  // For example, if a user posts a new node then we want to disable the slave
+  // server for that user temporarily to allow the slave server to catch up.
+  // That way, that user will see their changes immediately while for other
+  // users we still get the benefits of having a slave server, just with slightly
+  // stale data.  Code that wants to disable the slave server should use the
+  // db_set_ignore_slave() function to set $_SESSION['ignore_slave_server'] to
+  // the timestamp after which the slave can be re-enabled.
+  if (isset($_SESSION['ignore_slave_server'])) {
+    if ($_SESSION['ignore_slave_server'] >= $_SERVER['REQUEST_TIME']) {
+      $slave = FALSE;
+    }
+    else {
+      unset($_SESSION['ignore_slave_server']);
+    }
+  }
+
   if (variable_get('dev_query', 0)) {
     list($usec, $sec) = explode(' ', microtime());
     $timer = (float)$usec + (float)$sec;
diff --git a/modules/aggregator/aggregator.admin.inc b/modules/aggregator/aggregator.admin.inc
index 221d0ab..ca879f9 100644
--- a/modules/aggregator/aggregator.admin.inc
+++ b/modules/aggregator/aggregator.admin.inc
@@ -26,7 +26,15 @@ function aggregator_view() {
   $header = array(t('Title'), t('Items'), t('Last update'), t('Next update'), array('data' => t('Operations'), 'colspan' => '3'));
   $rows = array();
   while ($feed = db_fetch_object($result)) {
-    $rows[] = array(l($feed->title, "aggregator/sources/$feed->fid"), format_plural($feed->items, '1 item', '@count items'), ($feed->checked ? t('@time ago', array('@time' => format_interval(time() - $feed->checked))) : t('never')), ($feed->checked ? t('%time left', array('%time' => format_interval($feed->checked + $feed->refresh - time()))) : t('never')), l(t('edit'), "admin/content/aggregator/edit/feed/$feed->fid"), l(t('remove items'), "admin/content/aggregator/remove/$feed->fid"), l(t('update items'), "admin/content/aggregator/update/$feed->fid"));
+    $rows[] = array(
+      l($feed->title, "aggregator/sources/$feed->fid"),
+      format_plural($feed->items, '1 item', '@count items'),
+      ($feed->checked ? t('@time ago', array('@time' => format_interval(time() - $feed->checked))) : t('never')),
+      ($feed->checked ? t('%time left', array('%time' => format_interval($feed->checked + $feed->refresh - time()))) : t('never')),
+      l(t('edit'), "admin/content/aggregator/edit/feed/$feed->fid"),
+      l(t('remove items'), "admin/content/aggregator/remove/$feed->fid"),
+      l(t('update items'), "admin/content/aggregator/update/$feed->fid", array('query' => array('token' => drupal_get_token("aggregator/update/$feed->fid")))),
+    );
   }
   $output .= theme('table', $header, $rows);
 
@@ -209,6 +217,9 @@ function aggregator_admin_remove_feed_submit($form, &$form_state) {
  *   An associative array describing the feed to be refreshed.
  */
 function aggregator_admin_refresh_feed($feed) {
+  if (!isset($_GET['token']) || !drupal_valid_token($_GET['token'], 'aggregator/update/' . $feed['fid'])) {
+    return drupal_access_denied();
+  }
   aggregator_refresh($feed);
   drupal_goto('admin/content/aggregator');
 }
diff --git a/modules/aggregator/aggregator.info b/modules/aggregator/aggregator.info
index 6deb9da..cd72471 100644
--- a/modules/aggregator/aggregator.info
+++ b/modules/aggregator/aggregator.info
@@ -3,9 +3,3 @@ description = "Aggregates syndicated content (RSS, RDF, and Atom feeds)."
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/block/block.info b/modules/block/block.info
index 4a94c53..7e92e25 100644
--- a/modules/block/block.info
+++ b/modules/block/block.info
@@ -3,9 +3,3 @@ description = Controls the boxes that are displayed around the main content.
 package = Core - required
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/blog/blog.info b/modules/blog/blog.info
index c453dbb..a35fd27 100644
--- a/modules/blog/blog.info
+++ b/modules/blog/blog.info
@@ -3,9 +3,3 @@ description = Enables keeping easily and regularly updated user web pages or blo
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/blogapi/blogapi.info b/modules/blogapi/blogapi.info
index 1197eac..8abffbe 100644
--- a/modules/blogapi/blogapi.info
+++ b/modules/blogapi/blogapi.info
@@ -3,9 +3,3 @@ description = Allows users to post content using applications that support XML-R
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/book/book.info b/modules/book/book.info
index bce254b..999d68b 100644
--- a/modules/book/book.info
+++ b/modules/book/book.info
@@ -3,9 +3,3 @@ description = Allows users to structure site pages in a hierarchy or outline.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/color/color.info b/modules/color/color.info
index dc62a80..4e8ac66 100644
--- a/modules/color/color.info
+++ b/modules/color/color.info
@@ -3,9 +3,3 @@ description = Allows the user to change the color scheme of certain themes.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/comment/comment.info b/modules/comment/comment.info
index 0979541..983c5a8 100644
--- a/modules/comment/comment.info
+++ b/modules/comment/comment.info
@@ -3,9 +3,3 @@ description = Allows users to comment on and discuss published content.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/comment/comment.module b/modules/comment/comment.module
index e76f235..4a41101 100644
--- a/modules/comment/comment.module
+++ b/modules/comment/comment.module
@@ -774,6 +774,10 @@ function comment_save($edit) {
       // Clear the cache so an anonymous user can see his comment being added.
       cache_clear_all();
 
+      // Ignore slave server temporarily to give time for the
+      // saved comment to be propagated to the slave.
+      db_ignore_slave();
+
       // Explain the approval queue if necessary, and then
       // redirect the user to the node he's commenting on.
       if ($edit['status'] == COMMENT_NOT_PUBLISHED) {
diff --git a/modules/contact/contact.info b/modules/contact/contact.info
index d83c889..65693d0 100644
--- a/modules/contact/contact.info
+++ b/modules/contact/contact.info
@@ -3,9 +3,3 @@ description = Enables the use of both personal and site-wide contact forms.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/dblog/dblog.info b/modules/dblog/dblog.info
index 8326e75..af5d0d9 100644
--- a/modules/dblog/dblog.info
+++ b/modules/dblog/dblog.info
@@ -3,9 +3,3 @@ description = Logs and records system events to the database.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/filter/filter.info b/modules/filter/filter.info
index cfc59f4..3201bb1 100644
--- a/modules/filter/filter.info
+++ b/modules/filter/filter.info
@@ -3,9 +3,3 @@ description = Handles the filtering of content in preparation for display.
 package = Core - required
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/forum/forum.info b/modules/forum/forum.info
index 46f7f0a..c7e6d21 100644
--- a/modules/forum/forum.info
+++ b/modules/forum/forum.info
@@ -5,9 +5,3 @@ dependencies[] = comment
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/help/help.info b/modules/help/help.info
index 346c6ba..21c1f9e 100644
--- a/modules/help/help.info
+++ b/modules/help/help.info
@@ -3,9 +3,3 @@ description = Manages the display of online help.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/locale/locale.info b/modules/locale/locale.info
index fefef2b..fb2d9a5 100644
--- a/modules/locale/locale.info
+++ b/modules/locale/locale.info
@@ -3,9 +3,3 @@ description = Adds language handling functionality and enables the translation o
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/menu/menu.info b/modules/menu/menu.info
index 3b58654..4f21c69 100644
--- a/modules/menu/menu.info
+++ b/modules/menu/menu.info
@@ -3,9 +3,3 @@ description = Allows administrators to customize the site navigation menu.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/node/node.info b/modules/node/node.info
index ef05a2b..4bd95ef 100644
--- a/modules/node/node.info
+++ b/modules/node/node.info
@@ -3,9 +3,3 @@ description = Allows content to be submitted to the site and displayed on pages.
 package = Core - required
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/node/node.module b/modules/node/node.module
index 4842e43..c4efadb 100644
--- a/modules/node/node.module
+++ b/modules/node/node.module
@@ -941,6 +941,10 @@ function node_save(&$node) {
 
   // Clear the page and block caches.
   cache_clear_all();
+
+  // Ignore slave server temporarily to give time for the
+  // saved node to be propagated to the slave.
+  db_ignore_slave();
 }
 
 /**
diff --git a/modules/openid/openid.info b/modules/openid/openid.info
index 3c3dcea..0c49491 100644
--- a/modules/openid/openid.info
+++ b/modules/openid/openid.info
@@ -3,9 +3,3 @@ description = "Allows users to log into your site using OpenID."
 version = VERSION
 package = Core - optional
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/openid/openid.module b/modules/openid/openid.module
index b803d9d..0a07068 100644
--- a/modules/openid/openid.module
+++ b/modules/openid/openid.module
@@ -427,14 +427,17 @@ function openid_authentication($response) {
   elseif (variable_get('user_register', 1)) {
     // Register new user
     $form_state['redirect'] = NULL;
-    $form_state['values']['name'] = (empty($response['openid.sreg.nickname'])) ? '' : $response['openid.sreg.nickname'];
-    $form_state['values']['mail'] = (empty($response['openid.sreg.email'])) ? '' : $response['openid.sreg.email'];
+    // Only signed SREG keys are included as required by OpenID Simple
+    // Registration Extension 1.0, section 4.
+    $signed_keys = explode(',', $response['openid.signed']);
+    $form_state['values']['name'] = in_array('sreg.nickname', $signed_keys) ? $response['openid.sreg.nickname'] : '';
+    $form_state['values']['mail'] = in_array('sreg.email', $signed_keys) ? $response['openid.sreg.email'] : '';
     $form_state['values']['pass']  = user_password();
     $form_state['values']['status'] = variable_get('user_register', 1) == 1;
     $form_state['values']['response'] = $response;
     $form_state['values']['auth_openid'] = $identity;
 
-    if (empty($response['openid.sreg.email']) && empty($response['openid.sreg.nickname'])) {
+    if (empty($form_state['values']['name']) && empty($form_state['values']['mail'])) {
       drupal_set_message(t('Please complete the registration by filling out the form below. If you already have an account, you can <a href="@login">log in</a> now and add your OpenID under "My account".', array('@login' => url('user/login'))), 'warning');
       $success = FALSE;
     }
diff --git a/modules/path/path.info b/modules/path/path.info
index b34214c..8b9c52a 100644
--- a/modules/path/path.info
+++ b/modules/path/path.info
@@ -3,9 +3,3 @@ description = Allows users to rename URLs.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/php/php.info b/modules/php/php.info
index cdd83f1..558d129 100644
--- a/modules/php/php.info
+++ b/modules/php/php.info
@@ -3,9 +3,3 @@ description = Allows embedded PHP code/snippets to be evaluated.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/ping/ping.info b/modules/ping/ping.info
index 1fdd90f..c427829 100644
--- a/modules/ping/ping.info
+++ b/modules/ping/ping.info
@@ -3,9 +3,3 @@ description = Alerts other sites when your site has been updated.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/poll/poll.info b/modules/poll/poll.info
index 89d73ac..8edb999 100644
--- a/modules/poll/poll.info
+++ b/modules/poll/poll.info
@@ -3,9 +3,3 @@ description = Allows your site to capture votes on different topics in the form
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/profile/profile.info b/modules/profile/profile.info
index 607cc4a..a6764f5 100644
--- a/modules/profile/profile.info
+++ b/modules/profile/profile.info
@@ -3,9 +3,3 @@ description = Supports configurable user profiles.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/search/search.info b/modules/search/search.info
index bbfb6d4..5908fc0 100644
--- a/modules/search/search.info
+++ b/modules/search/search.info
@@ -3,9 +3,3 @@ description = Enables site-wide keyword searching.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/statistics/statistics.info b/modules/statistics/statistics.info
index 2969d87..b297368 100644
--- a/modules/statistics/statistics.info
+++ b/modules/statistics/statistics.info
@@ -3,9 +3,3 @@ description = Logs access statistics for your site.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/syslog/syslog.info b/modules/syslog/syslog.info
index 560c73f..81cda0a 100644
--- a/modules/syslog/syslog.info
+++ b/modules/syslog/syslog.info
@@ -3,9 +3,3 @@ description = Logs and records system events to syslog.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/system/system.info b/modules/system/system.info
index 47292aa..55986f9 100644
--- a/modules/system/system.info
+++ b/modules/system/system.info
@@ -3,9 +3,3 @@ description = Handles general site configuration for administrators.
 package = Core - required
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/system/system.module b/modules/system/system.module
index 5282418..4ca6c8e 100644
--- a/modules/system/system.module
+++ b/modules/system/system.module
@@ -8,7 +8,7 @@
 /**
  * The current system version.
  */
-define('VERSION', '6.22');
+define('VERSION', '6.23');
 
 /**
  * Core API compatibility.
diff --git a/modules/taxonomy/taxonomy.info b/modules/taxonomy/taxonomy.info
index 1bcb759..35900ec 100644
--- a/modules/taxonomy/taxonomy.info
+++ b/modules/taxonomy/taxonomy.info
@@ -3,9 +3,3 @@ description = Enables the categorization of content.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/throttle/throttle.info b/modules/throttle/throttle.info
index f4f3eb7..c1ed4e3 100644
--- a/modules/throttle/throttle.info
+++ b/modules/throttle/throttle.info
@@ -3,9 +3,3 @@ description = Handles the auto-throttling mechanism, to control site congestion.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/tracker/tracker.info b/modules/tracker/tracker.info
index ea4cdad..497bc94 100644
--- a/modules/tracker/tracker.info
+++ b/modules/tracker/tracker.info
@@ -4,9 +4,3 @@ dependencies[] = comment
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/translation/translation.info b/modules/translation/translation.info
index 776a206..d67587d 100644
--- a/modules/translation/translation.info
+++ b/modules/translation/translation.info
@@ -4,9 +4,3 @@ dependencies[] = locale
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/trigger/trigger.info b/modules/trigger/trigger.info
index a662ba0..fb6cebd 100644
--- a/modules/trigger/trigger.info
+++ b/modules/trigger/trigger.info
@@ -3,9 +3,3 @@ description = Enables actions to be fired on certain system events, such as when
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/update/update.info b/modules/update/update.info
index d92da58..aba6e82 100644
--- a/modules/update/update.info
+++ b/modules/update/update.info
@@ -3,9 +3,3 @@ description = Checks the status of available updates for Drupal and your install
 version = VERSION
 package = Core - optional
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/upload/upload.info b/modules/upload/upload.info
index 76732ca..32bed34 100644
--- a/modules/upload/upload.info
+++ b/modules/upload/upload.info
@@ -3,9 +3,3 @@ description = Allows users to upload and attach files to content.
 package = Core - optional
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/modules/user/user.info b/modules/user/user.info
index 2f0d99e..6df834c 100644
--- a/modules/user/user.info
+++ b/modules/user/user.info
@@ -3,9 +3,3 @@ description = Manages the user registration and login system.
 package = Core - required
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/bluemarine/bluemarine.info b/themes/bluemarine/bluemarine.info
index 1997bec..4182d98 100644
--- a/themes/bluemarine/bluemarine.info
+++ b/themes/bluemarine/bluemarine.info
@@ -3,9 +3,3 @@ description = Table-based multi-column theme with a marine and ash color scheme.
 version = VERSION
 core = 6.x
 engine = phptemplate
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/chameleon/chameleon.info b/themes/chameleon/chameleon.info
index e077d2d..5f9854b 100644
--- a/themes/chameleon/chameleon.info
+++ b/themes/chameleon/chameleon.info
@@ -10,9 +10,3 @@ stylesheets[all][] = style.css
 stylesheets[all][] = common.css
 version = VERSION
 core = 6.x
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/chameleon/marvin/marvin.info b/themes/chameleon/marvin/marvin.info
index 138290b..6685030 100644
--- a/themes/chameleon/marvin/marvin.info
+++ b/themes/chameleon/marvin/marvin.info
@@ -5,9 +5,3 @@ regions[right] = Right sidebar
 version = VERSION
 core = 6.x
 base theme = chameleon
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/garland/garland.info b/themes/garland/garland.info
index 5539987..b435812 100644
--- a/themes/garland/garland.info
+++ b/themes/garland/garland.info
@@ -5,9 +5,3 @@ core = 6.x
 engine = phptemplate
 stylesheets[all][] = style.css
 stylesheets[print][] = print.css
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/garland/minnelli/minnelli.info b/themes/garland/minnelli/minnelli.info
index 8f2bf30..21a4c38 100644
--- a/themes/garland/minnelli/minnelli.info
+++ b/themes/garland/minnelli/minnelli.info
@@ -4,9 +4,3 @@ version = VERSION
 core = 6.x
 base theme = garland
 stylesheets[all][] = minnelli.css
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
diff --git a/themes/pushbutton/pushbutton.info b/themes/pushbutton/pushbutton.info
index 1bd3d7d..5bd8f57 100644
--- a/themes/pushbutton/pushbutton.info
+++ b/themes/pushbutton/pushbutton.info
@@ -3,9 +3,3 @@ description = Tabled, multi-column theme in blue and orange tones.
 version = VERSION
 core = 6.x
 engine = phptemplate
-
-; Information added by drupal.org packaging script on 2011-05-25
-version = "6.22"
-project = "drupal"
-datestamp = "1306357015"
-
